{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "NeuroEvolution-clasificar-datos.ipynb",
      "provenance": [],
      "collapsed_sections": [],
      "toc_visible": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    },
    "accelerator": "GPU"
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "1Tu5MvS864Vv"
      },
      "source": [
        "#Demo NeuroEvolution usando Tensorflow-Neuroevolution framework (TFNE) para generar una RNA que pueda aprender a clasificar en base a ejemplos\n",
        "Fuente TFNE: https://github.com/PaulPauls/Tensorflow-Neuroevolution"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ZyCmxopcGYCB"
      },
      "source": [
        "0) Preparar ambiente:"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "cellView": "form",
        "id": "i79ux8zd64wv",
        "outputId": "d36dcc86-6366-4aaa-fc4c-fac129827b5a"
      },
      "source": [
        "#@title Instalar TFNE\n",
        "!pip install tfne\n",
        "\n",
        "###!git clone https://github.com/PaulPauls/Tensorflow-Neuroevolution\n",
        "###%cd Tensorflow-Neuroevolution\n",
        "###!pip install -r requirements.txt"
      ],
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: tfne in /usr/local/lib/python3.7/dist-packages (0.21.1)\n",
            "Requirement already satisfied: matplotlib in /usr/local/lib/python3.7/dist-packages (from tfne) (3.2.2)\n",
            "Requirement already satisfied: tensorflow>=2.0.0 in /usr/local/lib/python3.7/dist-packages (from tfne) (2.6.0)\n",
            "Requirement already satisfied: graphviz in /usr/local/lib/python3.7/dist-packages (from tfne) (0.10.1)\n",
            "Requirement already satisfied: PyQt5 in /usr/local/lib/python3.7/dist-packages (from tfne) (5.15.4)\n",
            "Requirement already satisfied: ray in /usr/local/lib/python3.7/dist-packages (from tfne) (1.6.0)\n",
            "Requirement already satisfied: pydot in /usr/local/lib/python3.7/dist-packages (from tfne) (1.3.0)\n",
            "Requirement already satisfied: typing-extensions~=3.7.4 in /usr/local/lib/python3.7/dist-packages (from tensorflow>=2.0.0->tfne) (3.7.4.3)\n",
            "Requirement already satisfied: h5py~=3.1.0 in /usr/local/lib/python3.7/dist-packages (from tensorflow>=2.0.0->tfne) (3.1.0)\n",
            "Requirement already satisfied: keras~=2.6 in /usr/local/lib/python3.7/dist-packages (from tensorflow>=2.0.0->tfne) (2.6.0)\n",
            "Requirement already satisfied: tensorflow-estimator~=2.6 in /usr/local/lib/python3.7/dist-packages (from tensorflow>=2.0.0->tfne) (2.6.0)\n",
            "Requirement already satisfied: protobuf>=3.9.2 in /usr/local/lib/python3.7/dist-packages (from tensorflow>=2.0.0->tfne) (3.17.3)\n",
            "Requirement already satisfied: gast==0.4.0 in /usr/local/lib/python3.7/dist-packages (from tensorflow>=2.0.0->tfne) (0.4.0)\n",
            "Requirement already satisfied: numpy~=1.19.2 in /usr/local/lib/python3.7/dist-packages (from tensorflow>=2.0.0->tfne) (1.19.5)\n",
            "Requirement already satisfied: flatbuffers~=1.12.0 in /usr/local/lib/python3.7/dist-packages (from tensorflow>=2.0.0->tfne) (1.12)\n",
            "Requirement already satisfied: wheel~=0.35 in /usr/local/lib/python3.7/dist-packages (from tensorflow>=2.0.0->tfne) (0.37.0)\n",
            "Requirement already satisfied: keras-preprocessing~=1.1.2 in /usr/local/lib/python3.7/dist-packages (from tensorflow>=2.0.0->tfne) (1.1.2)\n",
            "Requirement already satisfied: astunparse~=1.6.3 in /usr/local/lib/python3.7/dist-packages (from tensorflow>=2.0.0->tfne) (1.6.3)\n",
            "Requirement already satisfied: clang~=5.0 in /usr/local/lib/python3.7/dist-packages (from tensorflow>=2.0.0->tfne) (5.0)\n",
            "Requirement already satisfied: google-pasta~=0.2 in /usr/local/lib/python3.7/dist-packages (from tensorflow>=2.0.0->tfne) (0.2.0)\n",
            "Requirement already satisfied: six~=1.15.0 in /usr/local/lib/python3.7/dist-packages (from tensorflow>=2.0.0->tfne) (1.15.0)\n",
            "Requirement already satisfied: wrapt~=1.12.1 in /usr/local/lib/python3.7/dist-packages (from tensorflow>=2.0.0->tfne) (1.12.1)\n",
            "Requirement already satisfied: tensorboard~=2.6 in /usr/local/lib/python3.7/dist-packages (from tensorflow>=2.0.0->tfne) (2.6.0)\n",
            "Requirement already satisfied: opt-einsum~=3.3.0 in /usr/local/lib/python3.7/dist-packages (from tensorflow>=2.0.0->tfne) (3.3.0)\n",
            "Requirement already satisfied: absl-py~=0.10 in /usr/local/lib/python3.7/dist-packages (from tensorflow>=2.0.0->tfne) (0.12.0)\n",
            "Requirement already satisfied: termcolor~=1.1.0 in /usr/local/lib/python3.7/dist-packages (from tensorflow>=2.0.0->tfne) (1.1.0)\n",
            "Requirement already satisfied: grpcio<2.0,>=1.37.0 in /usr/local/lib/python3.7/dist-packages (from tensorflow>=2.0.0->tfne) (1.40.0)\n",
            "Requirement already satisfied: cached-property in /usr/local/lib/python3.7/dist-packages (from h5py~=3.1.0->tensorflow>=2.0.0->tfne) (1.5.2)\n",
            "Requirement already satisfied: markdown>=2.6.8 in /usr/local/lib/python3.7/dist-packages (from tensorboard~=2.6->tensorflow>=2.0.0->tfne) (3.3.4)\n",
            "Requirement already satisfied: tensorboard-data-server<0.7.0,>=0.6.0 in /usr/local/lib/python3.7/dist-packages (from tensorboard~=2.6->tensorflow>=2.0.0->tfne) (0.6.1)\n",
            "Requirement already satisfied: tensorboard-plugin-wit>=1.6.0 in /usr/local/lib/python3.7/dist-packages (from tensorboard~=2.6->tensorflow>=2.0.0->tfne) (1.8.0)\n",
            "Requirement already satisfied: google-auth<2,>=1.6.3 in /usr/local/lib/python3.7/dist-packages (from tensorboard~=2.6->tensorflow>=2.0.0->tfne) (1.35.0)\n",
            "Requirement already satisfied: setuptools>=41.0.0 in /usr/local/lib/python3.7/dist-packages (from tensorboard~=2.6->tensorflow>=2.0.0->tfne) (57.4.0)\n",
            "Requirement already satisfied: google-auth-oauthlib<0.5,>=0.4.1 in /usr/local/lib/python3.7/dist-packages (from tensorboard~=2.6->tensorflow>=2.0.0->tfne) (0.4.6)\n",
            "Requirement already satisfied: werkzeug>=0.11.15 in /usr/local/lib/python3.7/dist-packages (from tensorboard~=2.6->tensorflow>=2.0.0->tfne) (1.0.1)\n",
            "Requirement already satisfied: requests<3,>=2.21.0 in /usr/local/lib/python3.7/dist-packages (from tensorboard~=2.6->tensorflow>=2.0.0->tfne) (2.23.0)\n",
            "Requirement already satisfied: cachetools<5.0,>=2.0.0 in /usr/local/lib/python3.7/dist-packages (from google-auth<2,>=1.6.3->tensorboard~=2.6->tensorflow>=2.0.0->tfne) (4.2.2)\n",
            "Requirement already satisfied: pyasn1-modules>=0.2.1 in /usr/local/lib/python3.7/dist-packages (from google-auth<2,>=1.6.3->tensorboard~=2.6->tensorflow>=2.0.0->tfne) (0.2.8)\n",
            "Requirement already satisfied: rsa<5,>=3.1.4 in /usr/local/lib/python3.7/dist-packages (from google-auth<2,>=1.6.3->tensorboard~=2.6->tensorflow>=2.0.0->tfne) (4.7.2)\n",
            "Requirement already satisfied: requests-oauthlib>=0.7.0 in /usr/local/lib/python3.7/dist-packages (from google-auth-oauthlib<0.5,>=0.4.1->tensorboard~=2.6->tensorflow>=2.0.0->tfne) (1.3.0)\n",
            "Requirement already satisfied: importlib-metadata in /usr/local/lib/python3.7/dist-packages (from markdown>=2.6.8->tensorboard~=2.6->tensorflow>=2.0.0->tfne) (4.8.1)\n",
            "Requirement already satisfied: pyasn1<0.5.0,>=0.4.6 in /usr/local/lib/python3.7/dist-packages (from pyasn1-modules>=0.2.1->google-auth<2,>=1.6.3->tensorboard~=2.6->tensorflow>=2.0.0->tfne) (0.4.8)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.7/dist-packages (from requests<3,>=2.21.0->tensorboard~=2.6->tensorflow>=2.0.0->tfne) (2021.5.30)\n",
            "Requirement already satisfied: chardet<4,>=3.0.2 in /usr/local/lib/python3.7/dist-packages (from requests<3,>=2.21.0->tensorboard~=2.6->tensorflow>=2.0.0->tfne) (3.0.4)\n",
            "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/local/lib/python3.7/dist-packages (from requests<3,>=2.21.0->tensorboard~=2.6->tensorflow>=2.0.0->tfne) (1.24.3)\n",
            "Requirement already satisfied: idna<3,>=2.5 in /usr/local/lib/python3.7/dist-packages (from requests<3,>=2.21.0->tensorboard~=2.6->tensorflow>=2.0.0->tfne) (2.10)\n",
            "Requirement already satisfied: oauthlib>=3.0.0 in /usr/local/lib/python3.7/dist-packages (from requests-oauthlib>=0.7.0->google-auth-oauthlib<0.5,>=0.4.1->tensorboard~=2.6->tensorflow>=2.0.0->tfne) (3.1.1)\n",
            "Requirement already satisfied: zipp>=0.5 in /usr/local/lib/python3.7/dist-packages (from importlib-metadata->markdown>=2.6.8->tensorboard~=2.6->tensorflow>=2.0.0->tfne) (3.5.0)\n",
            "Requirement already satisfied: python-dateutil>=2.1 in /usr/local/lib/python3.7/dist-packages (from matplotlib->tfne) (2.8.2)\n",
            "Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.7/dist-packages (from matplotlib->tfne) (1.3.2)\n",
            "Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.1 in /usr/local/lib/python3.7/dist-packages (from matplotlib->tfne) (2.4.7)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.7/dist-packages (from matplotlib->tfne) (0.10.0)\n",
            "Requirement already satisfied: PyQt5-sip<13,>=12.8 in /usr/local/lib/python3.7/dist-packages (from PyQt5->tfne) (12.9.0)\n",
            "Requirement already satisfied: PyQt5-Qt5>=5.15 in /usr/local/lib/python3.7/dist-packages (from PyQt5->tfne) (5.15.2)\n",
            "Requirement already satisfied: attrs in /usr/local/lib/python3.7/dist-packages (from ray->tfne) (21.2.0)\n",
            "Requirement already satisfied: redis>=3.5.0 in /usr/local/lib/python3.7/dist-packages (from ray->tfne) (3.5.3)\n",
            "Requirement already satisfied: msgpack<2.0.0,>=1.0.0 in /usr/local/lib/python3.7/dist-packages (from ray->tfne) (1.0.2)\n",
            "Requirement already satisfied: pyyaml in /usr/local/lib/python3.7/dist-packages (from ray->tfne) (3.13)\n",
            "Requirement already satisfied: click>=7.0 in /usr/local/lib/python3.7/dist-packages (from ray->tfne) (7.1.2)\n",
            "Requirement already satisfied: filelock in /usr/local/lib/python3.7/dist-packages (from ray->tfne) (3.0.12)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "soqbTJnR-vgh"
      },
      "source": [
        "1) Cargar librerías:"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "b4oaPcXw-uzT",
        "cellView": "form",
        "outputId": "2afda7ff-b6f7-44fd-ed76-f40800e12d27"
      },
      "source": [
        "#@title Librerías a usar\n",
        "from absl import app, flags, logging\n",
        "from __future__ import annotations\n",
        "\n",
        "import numpy as np\n",
        "import tensorflow as tf\n",
        "import pandas as pd\n",
        "\n",
        "import os\n",
        "import csv\n",
        "\n",
        "import tfne\n",
        "from tfne.environments import BaseEnvironment\n",
        "from tfne.helper_functions import read_option_from_config\n",
        "\n",
        "from sklearn.model_selection import train_test_split\n",
        "from keras.utils import np_utils\n",
        "\n",
        "from sklearn.metrics import classification_report\n",
        "from sklearn.metrics import confusion_matrix\n",
        "\n",
        "print(\"Librerías cargadas\")"
      ],
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Librerías cargadas\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "cnL_qtVxM_DN"
      },
      "source": [
        "2) Determinar configuración:"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "yU_Vllt7GzIW",
        "cellView": "form"
      },
      "source": [
        "#@title Parámetros:\n",
        "\n",
        "#@markdown Criterio de Paro:\n",
        "ce_maximo_generaciones_procesar = 15 #@param {type:\"integer\"}\n",
        "ce_finalizar_al_encontrar_max_fitness = True #@param {type:\"boolean\"}\n",
        "#@markdown Función de Aptitud:\n",
        "fitness_calc_usando_exactitud = \"Suma Entrenamiento & Validacion\" #@param [\"Entrenamiento\", \"Validacion\", \"Min Entrenamiento & Validacion\", \"Promedio Entrenamiento & Validacion\", \"Suma Entrenamiento & Validacion\"]\n",
        "fitness_penalizar_x_topologia = True #@param {type:\"boolean\"}\n",
        "#@markdown Blueprints (topología & optimizador):\n",
        "ce_cant_poblacion_blueprints =  9#@param {type:\"integer\"}\n",
        "#@markdown  Genomas (instancias de RNAs):\n",
        "ce_cant_genomas_por_blueprint =  8#@param {type:\"integer\"}\n",
        "#@markdown Módulos (capas RNA):\n",
        "ce_cant_poblacion_modulos = 25 #@param {type:\"integer\"}\n",
        "ce_tipo_generacion_modulos = \"Dinamica\" #@param [\"Basica\", \"Dinamica\", \"Fija\"]\n",
        "#@markdown Operadores Genéticos:\n",
        "ce_max_mutacion = 0.3 #@param {type:\"number\"}\n",
        "ce_probab_mutacion = 0.4 #@param {type:\"number\"}\n",
        "ce_probab_cruzamiento = 0.6 #@param {type:\"number\"}\n",
        "#@markdown Entrenamiento RNA:\n",
        "rna_cant_epocas_entrenamiento = 30#@param {type:\"integer\"}\n",
        "rna_cant_epocas_entrenamiento_incrementa_por_generacion = 15#@param {type:\"integer\"}\n",
        "rna_cant_epocas_reentrenamiento_best_final =   350#@param {type:\"integer\"}\n",
        "\n"
      ],
      "execution_count": 3,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "a8aPzMxQNCT2",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "cellView": "form",
        "outputId": "fafb0314-d0f7-495d-fc35-1bbeaaab6289"
      },
      "source": [
        "#@title Datos:\n",
        "\n",
        "# monta Google Drive:\n",
        "# Nota: la primera vez se debe confirmar el uso logueandose en \"Google Drive File Stream\" y obteniendo código de autentificación.\n",
        "from google.colab import drive\n",
        "drive.mount('/content/gdrive')\n",
        "\n",
        "# directorio local en Google Drive\n",
        "path = '/content/gdrive/My Drive/IA/demoML/datos/'  #@param {type:\"string\"}\n",
        "\n",
        "## selección de los parámetros \n",
        "\n",
        "def cargarNombreClases(path, archivo_datos, atributo_clase, nombre_clases):\n",
        "  # importa definición de la clase\n",
        "  arClasesFN = archivo_datos.split('.')[0] + '_nombreClases.txt'\n",
        "  if os.path.isfile( path + '/' + arClasesFN ):\n",
        "    with open( path + '/' + arClasesFN, mode='r') as csvfile:\n",
        "        r = csv.reader(csvfile, delimiter=',')\n",
        "        auxAtributo = r.__next__()\n",
        "        auxClases = r.__next__()\n",
        "    print('\\n> Definición de los valores discretos para la clase cargada de ' + arClasesFN +'.\\n')\n",
        "    return auxAtributo[0], ','.join(auxClases)\n",
        "  else:\n",
        "    return atributo_clase, nombre_clases\n",
        "\n",
        "#@markdown ### Archivo de datos a utilizar:\n",
        "archivo_datos = 'ANIMALES.csv'  #@param {type:\"string\"}\n",
        "#@markdown ### Nombre del atributo clase / objetivo:\n",
        "atributo_clase = '-' #@param {type:\"string\"}\n",
        "#@markdown ### Descripción de los valores del atributo clase / objetivo:  (nota: siempre debe comienzar en 0, por lo que si no tiene valor 0, agregar \"na\")\n",
        "nombre_clases = '-' #@param {type:\"string\"}\n",
        "#@markdown ### Porcentaje de datos para usar en el entrenamiento:\n",
        "proporcion_porcentaje_datos_entrenamiento =   70#@param {type:\"integer\"}\n",
        "\n",
        "## aplicación de los parámetros elegidos\n",
        "\n",
        "if atributo_clase == '' or  atributo_clase == '-':\n",
        "  # trata de obtener la configuración del archivo asociado\n",
        "  atributo_clase, nombre_clases = cargarNombreClases(path, archivo_datos, atributo_clase, nombre_clases)\n",
        "\n",
        "\n",
        "# define nombre atributo de CLASE para ejemplo IRIS\n",
        "ClassAttributeName = atributo_clase\n",
        "\n",
        "# define valores de clases para ejemplo IRIS\n",
        "CLASES = [ ]\n",
        "for val in nombre_clases.split(','):\n",
        "  CLASES.append( val )\n",
        "\n",
        "# determina la proporción a usar para entrenar y probar\n",
        "if proporcion_porcentaje_datos_entrenamiento>100:\n",
        "  propTrain = 1\n",
        "elif proporcion_porcentaje_datos_entrenamiento<1:\n",
        "  propTrain = 0.1\n",
        "else:\n",
        "  propTrain = proporcion_porcentaje_datos_entrenamiento/100\n",
        "\n",
        "print(\"Configuración definida de \", archivo_datos)\n",
        "print(\"Atributo clase: \", ClassAttributeName, \": \", CLASES)\n",
        "\n",
        "# configura para que muestre todas las columnas y filas\n",
        "pd.options.display.max_rows = 100\n",
        "pd.options.display.max_columns = 100\n",
        "\n",
        "# Carga los datos del CSV y muestra los primeros\n",
        "df = pd.read_csv(path + archivo_datos)\n",
        "print(\"\\n> Cabecera: \")\n",
        "print(df.head())\n",
        "print(\"\\n> Características: \")\n",
        "print(df.describe())\n",
        "\n",
        "# genera los datos solo con la clase para entrenar y probar\n",
        "Y = np.array(df.pop(ClassAttributeName))\n",
        "X = np.array(df)\n",
        "\n"
      ],
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount(\"/content/gdrive\", force_remount=True).\n",
            "\n",
            "> Definición de los valores discretos para la clase cargada de ANIMALES_nombreClases.txt.\n",
            "\n",
            "Configuración definida de  ANIMALES.csv\n",
            "Atributo clase:  TIPO :  ['na', 'MAMIFERO', 'AVE', 'REPTIL', 'PEZ', 'ANFIBIO', 'INSECTO', 'INVERTEBRADO']\n",
            "\n",
            "> Cabecera: \n",
            "   TIENE_PELO  TIENE_PLUMAS  PONE_HUEVOS  DA_LECHE  ES_AEREO  ES_ACUATICO  \\\n",
            "0           1             0            0         1         0            0   \n",
            "1           1             0            0         1         0            0   \n",
            "2           0             0            1         0         0            1   \n",
            "3           1             0            0         1         0            0   \n",
            "4           1             0            0         1         0            0   \n",
            "\n",
            "   ES_PREDADOR  TIENE_DIENTES  TIENE_COLUMNA  RESPIRA  ES_VENENOSO  \\\n",
            "0            1              1              1        1            0   \n",
            "1            0              1              1        1            0   \n",
            "2            1              1              1        0            0   \n",
            "3            1              1              1        1            0   \n",
            "4            0              1              1        1            0   \n",
            "\n",
            "   TIENE_ALETAS  CANT_PATAS  TIENE_COLA  ES_DOMESTICO  TIPO  \n",
            "0             0           4           0             0     1  \n",
            "1             0           4           1             0     1  \n",
            "2             1           0           1             0     4  \n",
            "3             0           4           1             0     1  \n",
            "4             0           4           1             1     1  \n",
            "\n",
            "> Características: \n",
            "       TIENE_PELO  TIENE_PLUMAS  PONE_HUEVOS  DA_LECHE   ES_AEREO  \\\n",
            "count   50.000000     50.000000    50.000000  50.00000  50.000000   \n",
            "mean     0.340000      0.160000     0.680000   0.30000   0.220000   \n",
            "std      0.478518      0.370328     0.471212   0.46291   0.418452   \n",
            "min      0.000000      0.000000     0.000000   0.00000   0.000000   \n",
            "25%      0.000000      0.000000     0.000000   0.00000   0.000000   \n",
            "50%      0.000000      0.000000     1.000000   0.00000   0.000000   \n",
            "75%      1.000000      0.000000     1.000000   1.00000   0.000000   \n",
            "max      1.000000      1.000000     1.000000   1.00000   1.000000   \n",
            "\n",
            "       ES_ACUATICO  ES_PREDADOR  TIENE_DIENTES  TIENE_COLUMNA    RESPIRA  \\\n",
            "count    50.000000    50.000000      50.000000      50.000000  50.000000   \n",
            "mean      0.440000     0.600000       0.520000       0.720000   0.780000   \n",
            "std       0.501427     0.494872       0.504672       0.453557   0.418452   \n",
            "min       0.000000     0.000000       0.000000       0.000000   0.000000   \n",
            "25%       0.000000     0.000000       0.000000       0.000000   1.000000   \n",
            "50%       0.000000     1.000000       1.000000       1.000000   1.000000   \n",
            "75%       1.000000     1.000000       1.000000       1.000000   1.000000   \n",
            "max       1.000000     1.000000       1.000000       1.000000   1.000000   \n",
            "\n",
            "       ES_VENENOSO  TIENE_ALETAS  CANT_PATAS  TIENE_COLA  ES_DOMESTICO  \\\n",
            "count    50.000000      50.00000   50.000000   50.000000      50.00000   \n",
            "mean      0.160000       0.14000    2.980000    0.580000       0.14000   \n",
            "std       0.370328       0.35051    2.272192    0.498569       0.35051   \n",
            "min       0.000000       0.00000    0.000000    0.000000       0.00000   \n",
            "25%       0.000000       0.00000    2.000000    0.000000       0.00000   \n",
            "50%       0.000000       0.00000    3.000000    1.000000       0.00000   \n",
            "75%       0.000000       0.00000    4.000000    1.000000       0.00000   \n",
            "max       1.000000       1.00000    8.000000    1.000000       1.00000   \n",
            "\n",
            "            TIPO  \n",
            "count  50.000000  \n",
            "mean    3.480000  \n",
            "std     2.296759  \n",
            "min     1.000000  \n",
            "25%     1.000000  \n",
            "50%     3.000000  \n",
            "75%     6.000000  \n",
            "max     7.000000  \n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "G96unKkBIPDW"
      },
      "source": [
        "2) Preparar para NeuroEvolution:"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "FCbZ1lpH_wgQ",
        "cellView": "form",
        "outputId": "aeb215f2-580f-4cf7-d8c8-ca322bdea330"
      },
      "source": [
        "#@title Crear Archivo Configuración \n",
        "# ver explicación en https://tfne.readthedocs.io/en/latest/codeepneat/codeepneat-config.html \n",
        "\n",
        "confFileName = './codeepneat_config.cfg'\n",
        "\n",
        "s = \"\"\n",
        "with open(confFileName, 'w') as f:\n",
        "    s = s + \"[EVALUATION]\\n\"\n",
        "    s = s + \"epochs        = \" + str(rna_cant_epocas_entrenamiento) + \"\\n\"\n",
        "    s = s + \"batch_size    = None\\n\"\n",
        "    s = s + \"preprocessing = None\\n\"\n",
        "    if rna_cant_epocas_entrenamiento_incrementa_por_generacion > 0:      \n",
        "      s = s + \"increase_epochs_every_n_genomes = \" + str(ce_cant_poblacion_blueprints*ce_cant_genomas_por_blueprint) + \"\\n\"\n",
        "      s = s + \"increase_epochs_add_epochs = \" + str(rna_cant_epocas_entrenamiento_incrementa_por_generacion) + \"\\n\"\n",
        "    else:\n",
        "      s = s + \"increase_epochs_every_n_genomes = 0\\n\"\n",
        "      s = s + \"increase_epochs_add_epochs = 0\\n\"\n",
        "    s = s + \"\\n\"\n",
        "    s = s + \"\\n\"\n",
        "    s = s + \"[FITNESS]\\n\"\n",
        "    if fitness_calc_usando_exactitud == \"Entrenamiento\":\n",
        "      s = s + \"calc_using_accuracy    = 'E'\\n\"    \n",
        "    elif fitness_calc_usando_exactitud == \"Validacion\":\n",
        "      s = s + \"calc_using_accuracy    = 'V'\\n\"    \n",
        "    elif fitness_calc_usando_exactitud == \"Min Entrenamiento & Validacion\":\n",
        "      s = s + \"calc_using_accuracy    = 'M'\\n\"    \n",
        "    elif fitness_calc_usando_exactitud == \"Promedio Entrenamiento & Validacion\":\n",
        "      s = s + \"calc_using_accuracy    = 'P'\\n\"    \n",
        "    else:\n",
        "      s = s + \"calc_using_accuracy    = 'S'\\n\"        \n",
        "    s = s + \"penalize_based_on_topology  = \" + str(fitness_penalizar_x_topologia) + \"\\n\"\n",
        "    s = s + \"\\n\"\n",
        "    s = s + \"\\n\"\n",
        "    s = s + \"[POPULATION]\\n\"\n",
        "    s = s + \"bp_pop_size    = \" + str(ce_cant_poblacion_blueprints) + \"\\n\"\n",
        "    s = s + \"mod_pop_size   = \" + str(ce_cant_poblacion_modulos) + \"\\n\"\n",
        "    s = s + \"genomes_per_bp = \" + str(ce_cant_genomas_por_blueprint ) + \"\\n\"\n",
        "    s = s + \"\\n\"\n",
        "    s = s + \"\\n\"\n",
        "    s = s + \"[GENOME]\\n\"\n",
        "    s = s + \"dtype                = 'float32'\\n\"\n",
        "    s = s + \"available_modules    = ['DenseDropout']\\n\"\n",
        "    s = s + \"available_optimizers = ['SGD','Adam']\\n\"\n",
        "    s = s + \"output_layers        = [{'class_name': 'Dense', 'config': {'units': \"+ str(len(CLASES)) + \", 'activation': 'softmax'}}] \\n\"\n",
        "    s = s + \"\\n\"\n",
        "    s = s + \"\\n\"\n",
        "    s = s + \"[MODULE_SPECIATION]\\n\"\n",
        "    if ce_tipo_generacion_modulos == \"Dinamica\":\n",
        "      # Dinamica      \n",
        "      s = s + \"mod_spec_type            = 'param-distance-dynamic'\\n\"\n",
        "      s = s + \"mod_spec_species_count   = 4\\n\"\n",
        "      s = s + \"mod_spec_distance        = 0.3\\n\"\n",
        "      s = s + \"mod_spec_mod_elitism     = 2\\n\"\n",
        "      s = s + \"mod_spec_min_offspring   = 1\\n\"\n",
        "      s = s + \"mod_spec_reprod_thres    = 0.5\\n\"\n",
        "      s = s + \"mod_spec_max_stagnation  = 10\\n\"\n",
        "      s = s + \"mod_spec_species_elitism = 2\\n\"\n",
        "      s = s + \"mod_spec_rebase_repr     = True\\n\"\n",
        "      s = s + \"mod_spec_reinit_extinct  = False\\n\"\n",
        "    elif ce_tipo_generacion_modulos == \"Fija\":\n",
        "      # Fija\n",
        "      s = s + \"mod_spec_type            = 'param-distance-fixed'\\n\"\n",
        "      s = s + \"mod_spec_distance        = 0.3\\n\"\n",
        "      s = s + \"mod_spec_mod_elitism     = 2\\n\"\n",
        "      s = s + \"mod_spec_min_offspring   = 1\\n\"\n",
        "      s = s + \"mod_spec_reprod_thres    = 0.5\\n\"\n",
        "      s = s + \"mod_spec_max_stagnation  = 10\\n\"\n",
        "      s = s + \"mod_spec_species_elitism = 2\\n\"\n",
        "      s = s + \"mod_spec_rebase_repr     = True\\n\"\n",
        "      s = s + \"mod_spec_reinit_extinct  = False\\n\"\n",
        "    else:\n",
        "      # Basica\n",
        "      s = s + \"mod_spec_type          = 'basic'\\n\"\n",
        "      s = s + \"mod_spec_mod_elitism   = 4\\n\"\n",
        "      s = s + \"mod_spec_min_offspring = 1\\n\"\n",
        "      s = s + \"mod_spec_reprod_thres  = 0.5\\n\"\n",
        "    s = s + \"\\n\"\n",
        "    s = s + \"\\n\"\n",
        "    s = s + \"[MODULE_EVOLUTION]\\n\"\n",
        "    s = s + \"mod_max_mutation   = \" + str(ce_max_mutacion) + \"\\n\"\n",
        "    s = s + \"mod_mutation_prob  = \" + str(ce_probab_mutacion) + \"\\n\"\n",
        "    s = s + \"mod_crossover_prob = \" + str(ce_probab_cruzamiento) + \"\\n\"\n",
        "    s = s + \"\\n\"\n",
        "    s = s + \"\\n\"\n",
        "    s = s + \"[BP_SPECIATION]\\n\"\n",
        "    if ce_tipo_generacion_modulos == \"Dinamica\":\n",
        "      # Dinamica\n",
        "      s = s + \"bp_spec_type            = 'gene-overlap-dynamic'\\n\"\n",
        "      s = s + \"bp_spec_species_count   = 3\\n\"\n",
        "      s = s + \"bp_spec_distance        = 0.3\\n\"\n",
        "      s = s + \"bp_spec_bp_elitism      = 2\\n\"\n",
        "      s = s + \"bp_spec_min_offspring   = 1\\n\"\n",
        "      s = s + \"bp_spec_reprod_thres    = 0.5\\n\"\n",
        "      s = s + \"bp_spec_max_stagnation  = 15\\n\"\n",
        "      s = s + \"bp_spec_species_elitism = 2\\n\"\n",
        "      s = s + \"bp_spec_rebase_repr     = True\\n\"\n",
        "      s = s + \"bp_spec_reinit_extinct  = True\\n\"\n",
        "    elif ce_tipo_generacion_modulos == \"Fija\":\n",
        "      # Fija\n",
        "      s = s + \"bp_spec_type            = 'gene-overlap-fixed'\\n\"\n",
        "      s = s + \"bp_spec_distance        = 0.3\\n\"\n",
        "      s = s + \"bp_spec_bp_elitism      = 2\\n\"\n",
        "      s = s + \"bp_spec_min_offspring   = 1\\n\"\n",
        "      s = s + \"bp_spec_reprod_thres    = 0.5\\n\"\n",
        "      s = s + \"bp_spec_max_stagnation  = 15\\n\"\n",
        "      s = s + \"bp_spec_species_elitism = 2\\n\"\n",
        "      s = s + \"bp_spec_rebase_repr     = True\\n\"\n",
        "      s = s + \"bp_spec_reinit_extinct  = True\\n\"  \n",
        "    else:\n",
        "      # Basica    \n",
        "      s = s + \"bp_spec_type          = 'basic'\\n\"\n",
        "      s = s + \"bp_spec_bp_elitism    = 2\\n\"\n",
        "      s = s + \"bp_spec_min_offspring = 1\\n\"\n",
        "      s = s + \"bp_spec_reprod_thres  = 0.5\\n\"\n",
        "    s = s + \"\\n\"\n",
        "    s = s + \"\\n\"\n",
        "    s = s + \"[BP_EVOLUTION]\\n\"    \n",
        "    if ce_tipo_generacion_modulos == \"Dinamica\":\n",
        "      # Dinamica      \n",
        "      s = s + \"bp_max_mutation            = 0.3\\n\"    \n",
        "      s = s + \"bp_mutation_add_conn_prob  = 0.2\\n\"\n",
        "      s = s + \"bp_mutation_add_node_prob  = 0.2\\n\"\n",
        "      s = s + \"bp_mutation_rem_conn_prob  = 0.05\\n\"\n",
        "      s = s + \"bp_mutation_rem_node_prob  = 0.05\\n\"\n",
        "      s = s + \"bp_mutation_node_spec_prob = 0.3\\n\"\n",
        "      s = s + \"bp_mutation_optimizer_prob = 0.1\\n\"\n",
        "      s = s + \"bp_crossover_prob          = 0.1\\n\"\n",
        "    else:\n",
        "      # Fija o Basica\n",
        "      s = s + \"bp_max_mutation            = 0.3\\n\"    \n",
        "      s = s + \"bp_mutation_add_conn_prob  = 0.3\\n\"\n",
        "      s = s + \"bp_mutation_add_node_prob  = 0.3\\n\"\n",
        "      s = s + \"bp_mutation_rem_conn_prob  = 0.05\\n\"\n",
        "      s = s + \"bp_mutation_rem_node_prob  = 0.05\\n\"\n",
        "      s = s + \"bp_mutation_node_spec_prob = 0.1\\n\"\n",
        "      s = s + \"bp_mutation_optimizer_prob = 0.1\\n\"\n",
        "      s = s + \"bp_crossover_prob          = 0.1\\n\"\n",
        "    s = s + \"\\n\"\n",
        "    s = s + \"\\n\"\n",
        "    s = s + \"[MODULE_DENSEDROPOUT]\\n\"\n",
        "    s = s + \"merge_method = [{'class_name': 'Concatenate', 'config': {'axis': -1}}]\\n\"\n",
        "    s = s + \"units        = {'min': 2, 'max': 100, 'step': 4, 'stddev': 6}\\n\"\n",
        "    s = s + \"activation   = ['linear', 'relu', 'sigmoid', 'softmax', 'tanh']\\n\"\n",
        "    s = s + \"kernel_init  = ['glorot_normal', 'he_normal']\\n\"\n",
        "    s = s + \"bias_init    = ['zeros']\\n\"\n",
        "    s = s + \"dropout_flag = 0.5\\n\"\n",
        "    s = s + \"dropout_rate = {'min': 0.1, 'max': 0.4, 'step': 0.1, 'stddev': 0.1}\\n\"\n",
        "    s = s + \"\\n\"\n",
        "    s = s + \"\\n\"\n",
        "    s = s + \"[OPTIMIZER_SGD]\\n\"\n",
        "    s = s + \"learning_rate = {'min': 0.1, 'max': 0.3, 'step': 0.05, 'stddev': 0.05}\\n\"\n",
        "    s = s + \"momentum      = {'min': 0.3, 'max': 0.7, 'step': 0.1, 'stddev': 0.1}\\n\"\n",
        "    s = s + \"nesterov      = [True, False] \\n\"\n",
        "    s = s + \"\\n\"\n",
        "    s = s + \"\\n\"\n",
        "    s = s + \"[OPTIMIZER_ADAM]\\n\"\n",
        "    s = s + \"learning_rate = {'min': 0.0001, 'max': 0.1, 'step': 0.0001, 'stddev': 0.02}\\n\"\n",
        "    s = s + \"beta_1        = {'min': 0.6, 'max': 1.5, 'step': 0.05, 'stddev': 0.2}\\n\"\n",
        "    s = s + \"beta_2        = {'min': 0.8, 'max': 1.2, 'step': 0.001, 'stddev': 0.1}\\n\"\n",
        "    s = s + \"epsilon       = {'min': 1e-8, 'max': 1e-6, 'step': 1e-8, 'stddev': 1e-7}\\n\"\n",
        "    s = s + \"\\n\"\n",
        "    s = s + \"\\n\"\n",
        "    \n",
        "    f.write(s)\n",
        "\n",
        "# muestra nuevo archivo modificado\n",
        "%cat {confFileName}"
      ],
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[EVALUATION]\n",
            "epochs        = 30\n",
            "batch_size    = None\n",
            "preprocessing = None\n",
            "increase_epochs_every_n_genomes = 72\n",
            "increase_epochs_add_epochs = 15\n",
            "\n",
            "\n",
            "[FITNESS]\n",
            "calc_using_accuracy    = 'S'\n",
            "penalize_based_on_topology  = True\n",
            "\n",
            "\n",
            "[POPULATION]\n",
            "bp_pop_size    = 9\n",
            "mod_pop_size   = 25\n",
            "genomes_per_bp = 8\n",
            "\n",
            "\n",
            "[GENOME]\n",
            "dtype                = 'float32'\n",
            "available_modules    = ['DenseDropout']\n",
            "available_optimizers = ['SGD','Adam']\n",
            "output_layers        = [{'class_name': 'Dense', 'config': {'units': 8, 'activation': 'softmax'}}] \n",
            "\n",
            "\n",
            "[MODULE_SPECIATION]\n",
            "mod_spec_type            = 'param-distance-dynamic'\n",
            "mod_spec_species_count   = 4\n",
            "mod_spec_distance        = 0.3\n",
            "mod_spec_mod_elitism     = 2\n",
            "mod_spec_min_offspring   = 1\n",
            "mod_spec_reprod_thres    = 0.5\n",
            "mod_spec_max_stagnation  = 10\n",
            "mod_spec_species_elitism = 2\n",
            "mod_spec_rebase_repr     = True\n",
            "mod_spec_reinit_extinct  = False\n",
            "\n",
            "\n",
            "[MODULE_EVOLUTION]\n",
            "mod_max_mutation   = 0.3\n",
            "mod_mutation_prob  = 0.4\n",
            "mod_crossover_prob = 0.6\n",
            "\n",
            "\n",
            "[BP_SPECIATION]\n",
            "bp_spec_type            = 'gene-overlap-dynamic'\n",
            "bp_spec_species_count   = 3\n",
            "bp_spec_distance        = 0.3\n",
            "bp_spec_bp_elitism      = 2\n",
            "bp_spec_min_offspring   = 1\n",
            "bp_spec_reprod_thres    = 0.5\n",
            "bp_spec_max_stagnation  = 15\n",
            "bp_spec_species_elitism = 2\n",
            "bp_spec_rebase_repr     = True\n",
            "bp_spec_reinit_extinct  = True\n",
            "\n",
            "\n",
            "[BP_EVOLUTION]\n",
            "bp_max_mutation            = 0.3\n",
            "bp_mutation_add_conn_prob  = 0.2\n",
            "bp_mutation_add_node_prob  = 0.2\n",
            "bp_mutation_rem_conn_prob  = 0.05\n",
            "bp_mutation_rem_node_prob  = 0.05\n",
            "bp_mutation_node_spec_prob = 0.3\n",
            "bp_mutation_optimizer_prob = 0.1\n",
            "bp_crossover_prob          = 0.1\n",
            "\n",
            "\n",
            "[MODULE_DENSEDROPOUT]\n",
            "merge_method = [{'class_name': 'Concatenate', 'config': {'axis': -1}}]\n",
            "units        = {'min': 2, 'max': 100, 'step': 4, 'stddev': 6}\n",
            "activation   = ['linear', 'relu', 'sigmoid', 'softmax', 'tanh']\n",
            "kernel_init  = ['glorot_normal', 'he_normal']\n",
            "bias_init    = ['zeros']\n",
            "dropout_flag = 0.5\n",
            "dropout_rate = {'min': 0.1, 'max': 0.4, 'step': 0.1, 'stddev': 0.1}\n",
            "\n",
            "\n",
            "[OPTIMIZER_SGD]\n",
            "learning_rate = {'min': 0.1, 'max': 0.3, 'step': 0.05, 'stddev': 0.05}\n",
            "momentum      = {'min': 0.3, 'max': 0.7, 'step': 0.1, 'stddev': 0.1}\n",
            "nesterov      = [True, False] \n",
            "\n",
            "\n",
            "[OPTIMIZER_ADAM]\n",
            "learning_rate = {'min': 0.0001, 'max': 0.1, 'step': 0.0001, 'stddev': 0.02}\n",
            "beta_1        = {'min': 0.6, 'max': 1.5, 'step': 0.05, 'stddev': 0.2}\n",
            "beta_2        = {'min': 0.8, 'max': 1.2, 'step': 0.001, 'stddev': 0.1}\n",
            "epsilon       = {'min': 1e-8, 'max': 1e-6, 'step': 1e-8, 'stddev': 1e-7}\n",
            "\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "OJDMMi5ien2n",
        "cellView": "form",
        "outputId": "d2abf34f-1125-48d0-f869-e29981f7e582"
      },
      "source": [
        "#@title Preparar datos\n",
        "\n",
        "# separa al azar usando muestreo estratificado con proporción indicada\n",
        "x_train, x_test, y_train, y_test = train_test_split(X, Y, test_size=(1-propTrain), stratify=Y)\n",
        "\n",
        "# muestra resultados\n",
        "print(\"> Para Ambos: \")\n",
        "print(\" - dictMapeo: \", CLASES)\n",
        "\n",
        "print(\"\\n> Para Entrenamiento: \")\n",
        "print(\" - x_train (cant ejemplos, datos entrada): \", x_train.shape)\n",
        "print(\" - y_train (cant): \", len(y_train))\n",
        "for i in range(len(CLASES)):\n",
        "  cant = 0\n",
        "  for y in y_train:\n",
        "    if i == y: cant = cant + 1\n",
        "  print(\"    \", CLASES[i], \"[\", i, \"]:\", cant)\n",
        "\n",
        "print(\"\\n Para Prueba: \")\n",
        "print(\" - x_test (cant ejemplos, datos entrada): \", x_test.shape)\n",
        "print(\" - y_test (cant): \", len(y_test))\n",
        "for i in range(len(CLASES)):\n",
        "  cant = 0\n",
        "  for y in y_test:\n",
        "    if i == y: cant = cant + 1\n",
        "  print(\"    \", CLASES[i], \"[\", i, \"]:\", cant)\n"
      ],
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "> Para Ambos: \n",
            " - dictMapeo:  ['na', 'MAMIFERO', 'AVE', 'REPTIL', 'PEZ', 'ANFIBIO', 'INSECTO', 'INVERTEBRADO']\n",
            "\n",
            "> Para Entrenamiento: \n",
            " - x_train (cant ejemplos, datos entrada):  (34, 15)\n",
            " - y_train (cant):  34\n",
            "     na [ 0 ]: 0\n",
            "     MAMIFERO [ 1 ]: 10\n",
            "     AVE [ 2 ]: 6\n",
            "     REPTIL [ 3 ]: 3\n",
            "     PEZ [ 4 ]: 3\n",
            "     ANFIBIO [ 5 ]: 3\n",
            "     INSECTO [ 6 ]: 4\n",
            "     INVERTEBRADO [ 7 ]: 5\n",
            "\n",
            " Para Prueba: \n",
            " - x_test (cant ejemplos, datos entrada):  (16, 15)\n",
            " - y_test (cant):  16\n",
            "     na [ 0 ]: 0\n",
            "     MAMIFERO [ 1 ]: 5\n",
            "     AVE [ 2 ]: 2\n",
            "     REPTIL [ 3 ]: 2\n",
            "     PEZ [ 4 ]: 1\n",
            "     ANFIBIO [ 5 ]: 1\n",
            "     INSECTO [ 6 ]: 2\n",
            "     INVERTEBRADO [ 7 ]: 3\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "6QNOESjoDj6B",
        "cellView": "code",
        "outputId": "2cce068b-4079-41ca-e670-659615adc896"
      },
      "source": [
        "#@title Definir clase para el ambiente RNAEnvironment\n",
        "\n",
        "class RNAEnvironment(BaseEnvironment):\n",
        "    \"\"\"\n",
        "    TFNE compatible environment for the RNA\n",
        "    \"\"\"\n",
        "\n",
        "    def __init__(self, weight_training, data_x, data_y, config=None, verbosity=0, **kwargs):\n",
        "        \"\"\"\n",
        "        Initializes environment by setting up the dataset and processing the supplied config or supplied config\n",
        "        parameters. The configuration of the environment can either be supplied via a config file or via seperate config\n",
        "        parameters in the initialization.\n",
        "        @param weight_training: bool flag, indicating wether evaluation should be weight training or not\n",
        "        @param config: ConfigParser instance holding an 'Environment' section specifying the required environment\n",
        "                       parameters for the chosen evaluation method.\n",
        "        @param verbosity: integer specifying the verbosity of the evaluation\n",
        "        @param kwargs: Optionally supplied dict of each configuration parameter seperately in order to allow the\n",
        "                       creation of the evaluation environment without the requirement of a config file.\n",
        "        \"\"\"\n",
        "        # Initialize corresponding input and output mappings\n",
        "        print(\"> Preparando el ambiente...\")\n",
        "\n",
        "        # Initialize loss function to evaluate performance on either evaluation method and safe verbosity parameter\n",
        "        self.accuracy_metric = tf.keras.metrics.Accuracy()\n",
        "        self.verbosity = verbosity\n",
        "\n",
        "        # Determine and setup explicit evaluation method in accordance to supplied parameters\n",
        "        if not weight_training:\n",
        "            # Set up environment as non-weight training, requiring no parameters\n",
        "            self.eval_genome_fitness = self._eval_genome_fitness_non_weight_training\n",
        "\n",
        "        elif config is None and len(kwargs) == 0:\n",
        "            raise RuntimeError(\"No se han definido los parámetros para poder realizar la evolución y el entrenamiento de las RNA\")\n",
        "\n",
        "        elif len(kwargs) == 0:\n",
        "            # Set up environment as weight training and with a supplied config file\n",
        "            self.eval_genome_fitness = self._eval_genome_fitness_weight_training\n",
        "            self.epochs = read_option_from_config(config, 'EVALUATION', 'epochs')\n",
        "            self.batch_size = read_option_from_config(config, 'EVALUATION', 'batch_size')\n",
        "            self.increase_epochs_every_n_genomes = read_option_from_config(config, 'EVALUATION', 'increase_epochs_every_n_genomes')\n",
        "            self.increase_epochs_add_epochs = read_option_from_config(config, 'EVALUATION', 'increase_epochs_add_epochs')\n",
        "            #  parámetros para calculo fitness\n",
        "            self.calcFitness = read_option_from_config(config, 'FITNESS', 'calc_using_accuracy')\n",
        "            self.penalFitness = read_option_from_config(config, 'FITNESS', 'penalize_based_on_topology')\n",
        "\n",
        "        elif config is None:\n",
        "            # Set up environment as weight training and explicitely supplied parameters\n",
        "            self.eval_genome_fitness = self._eval_genome_fitness_weight_training\n",
        "            self.epochs = kwargs['epochs']\n",
        "            self.batch_size = kwargs['batch_size']\n",
        "            self.increase_epochs_every_n_genomes = kwargs['increase_epochs_every_n_genomes']\n",
        "            self.increase_epochs_add_epochs = kwargs['increase_epochs_add_epochs']            \n",
        "            # determina parámetros para calculo fitness\n",
        "            self.calcFitness = kwargs['fitness_calc_using_accuracy']\n",
        "            self.penalFitness = kwargs['fitness_penalize_based_on_topology']\n",
        "\n",
        "        print(\" Definiendo parámetros para cálculo de aptitud: \")\n",
        "        print(\"        calcFitness = \", self.calcFitness)\n",
        "        print(\"        penalFitness = \", self.penalFitness)\n",
        "\n",
        "        if self.calcFitness == \"E\":\n",
        "            # se toman todos los datos para entrenamiento\n",
        "            x_t, x_v, y_t, y_v = data_x, [], data_y, []\n",
        "        else:\n",
        "            # separa al azar usando muestreo al azar del 10%\n",
        "            # para tomar algunos como datos de validación\n",
        "            x_t, x_v, y_t, y_v = train_test_split(data_x, \n",
        "                                                data_y, \n",
        "                                                test_size=0.1)\n",
        "\n",
        "        print(\"  Definiendo datos: de los \", len(x_train), \"ejemplos de entrenamiento: \")\n",
        "        print(\"                      se usan \", len(x_t), \"ejemplos para entrenar \")\n",
        "        print(\"                      y \", len(x_v), \"ejemplos para validar.\")\n",
        "        print(\"\\n\")\n",
        "        self.x_train = np.array(x_t)\n",
        "        self.y_train = np.array(y_t)\n",
        "\n",
        "        self.x_val = np.array(x_v)\n",
        "        self.y_val = np.array(y_v)\n",
        "\n",
        "        # genera salida codificada para entrenamiento con capa softMax\n",
        "        self.y_trainEnc = np_utils.to_categorical(self.y_train)\n",
        "\n",
        "        # determina neuronas de entrada y salida\n",
        "        self.cantX = self.x_train.shape[1]\n",
        "        self.cantY = self.y_trainEnc.shape[1]\n",
        "        return\n",
        "\n",
        "    def eval_genome_fitness(self, genome) -> float:\n",
        "        # TO BE OVERRIDEN\n",
        "        raise RuntimeError()\n",
        "  \n",
        "    def _calculate_fitness(self, model, penalFitness=False) -> float:\n",
        "        \"\"\"\n",
        "        The genomes fitness is then calculated and returned as\n",
        "        the percentage of training and/or validation examples classified correctly.           \n",
        "        \"\"\"\n",
        "        # calcula para datos de entrenamiento\n",
        "        if len(self.y_train)>0:\n",
        "          self.accuracy_metric.reset_states()\n",
        "          self.accuracy_metric.update_state(self.y_train, np.argmax(model(self.x_train), axis=-1))\n",
        "          evaluated_fitness_train = round(self.accuracy_metric.result().numpy() * 100, 4)        \n",
        "        else:\n",
        "          evaluated_fitness_train = 0.0\n",
        "        # calcula para datos de entrenamiento\n",
        "        if len(self.y_val)>0:\n",
        "          self.accuracy_metric.reset_states()\n",
        "          self.accuracy_metric.update_state(self.y_val, np.argmax(model(self.x_val), axis=-1))\n",
        "          evaluated_fitness_val = round(self.accuracy_metric.result().numpy() * 100, 4)        \n",
        "        else:\n",
        "          evaluated_fitness_val = 0.0\n",
        "        # calcula final como promedio         \n",
        "        if self.calcFitness == \"E\":\n",
        "          # sólo de entrenamiento\n",
        "          evaluated_fitness = evaluated_fitness_train\n",
        "        elif self.calcFitness == \"V\":\n",
        "          # sólo de validación\n",
        "          evaluated_fitness = evaluated_fitness_val\n",
        "        elif self.calcFitness == \"P\":\n",
        "          # promedio ente ambos\n",
        "          evaluated_fitness = (evaluated_fitness_train + evaluated_fitness_val)/2\n",
        "        elif self.calcFitness ==  \"S\":\n",
        "          # suma ente ambos\n",
        "          evaluated_fitness = (evaluated_fitness_train + evaluated_fitness_val)\n",
        "        else:\n",
        "          # mínimo ente ambos\n",
        "          evaluated_fitness = min([evaluated_fitness_train, evaluated_fitness_val])\n",
        "        if penalFitness:\n",
        "          # sólo se aplica la penalización\n",
        "          # si está cercano a alcanzar la aptitud máxima\n",
        "          maxFit = self.return_MaxFitness()\n",
        "          if evaluated_fitness >= (maxFit - 11):\n",
        "            # penaliza aptitud teniendo en cuenta la complejidad de la RNA          \n",
        "            penalLayers = -1.5\n",
        "            ajusteCantPesos = self.cantX\n",
        "            for i in range(len(model.layers)):\n",
        "              l = model.layers[i]\n",
        "              tipoLay = str(type(l))            \n",
        "              if \"Dense\" in tipoLay:                \n",
        "                if i > 1:\n",
        "                  # contabiliza en base a la cantidad de pesos de las conexiones\n",
        "                  # nota: no tiene en cuenta los pesos de la capa de entrada ni de salida\n",
        "                  cantPesos = len(l.get_weights()[0]) \n",
        "                  penalLayers = penalLayers + (cantPesos / ajusteCantPesos)\n",
        "              elif \"Dropout\" in tipoLay:\n",
        "                penalLayers = penalLayers + 0.8\n",
        "              elif \"Concatenate\" in tipoLay:\n",
        "                penalLayers = penalLayers + 0.9\n",
        "            if penalLayers > 0:\n",
        "              evaluated_fitness = evaluated_fitness + 10.0 - (penalLayers / 10.0)\n",
        "        return round(evaluated_fitness, 5)\n",
        "    \n",
        "    def return_MaxFitness(self) -> float:\n",
        "      # devuelve valor máximo posible para la aptitud\n",
        "      if self.calcFitness ==  \"S\":\n",
        "        valMaxFit = 200.0\n",
        "      else:\n",
        "        valMaxFit = 100.0\n",
        "      if self.penalFitness:\n",
        "        valMaxFit = valMaxFit + 10\n",
        "      return valMaxFit\n",
        "\n",
        "    def _eval_genome_fitness_weight_training(self, genome) -> float:\n",
        "        \"\"\"\n",
        "        Evaluates the genome's fitness by obtaining the associated Tensorflow model and optimizer, \n",
        "        compiling them and then training them for the config specified duration.      \n",
        "        @param genome: TFNE compatible genome that is to be evaluated\n",
        "        @return: genome calculated fitness\n",
        "        \"\"\"\n",
        "        # Get model and optimizer required for compilation\n",
        "        model = genome.get_model()\n",
        "        optimizer = genome.get_optimizer()\n",
        "\n",
        "        # si corresponde incrementa la cantidad de épocas\n",
        "        gen_id = genome.get_id()\n",
        "        if gen_id > self.increase_epochs_every_n_genomes:\n",
        "          cantEpochsIncr = int((gen_id / self.increase_epochs_every_n_genomes) * self.increase_epochs_add_epochs)\n",
        "          cant_epochs = self.epochs + cantEpochsIncr\n",
        "          #print(\"  > \", gen_id, self.epochs, cantEpochsIncr, cant_epochs)\n",
        "        else:\n",
        "          cant_epochs = self.epochs\n",
        "\n",
        "        # Compile and train model\n",
        "        model.compile(optimizer=optimizer, \n",
        "                      loss='categorical_crossentropy', metrics=['accuracy'])\n",
        "                      #loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True))        \n",
        "        model.fit(x=self.x_train, \n",
        "                  y=self.y_trainEnc, \n",
        "                  epochs=cant_epochs, \n",
        "                  batch_size=self.batch_size, \n",
        "                  verbose=self.verbosity)\n",
        "\n",
        "        # Evaluate and return its fitness\n",
        "        return self._calculate_fitness(model, self.penalFitness)\n",
        "\n",
        "    def _eval_genome_fitness_non_weight_training(self, genome) -> float:\n",
        "        raise NotImplementedError(\"Non-Weight training evaluation not yet implemented for Environment\")\n",
        "\n",
        "    def replay_genome(self, genome):\n",
        "        \"\"\"\n",
        "        Replay genome on environment by calculating its fitness and printing it.\n",
        "        @param genome: TFNE compatible genome that is to be evaluated\n",
        "        \"\"\"\n",
        "        print(\"> Probando Genome #{}:\".format(genome.get_id()))\n",
        "\n",
        "        # Determine fitness by creating model predictions with test images and then judging the fitness based on the\n",
        "        # achieved model accuracy.\n",
        "        model = genome.get_model()\n",
        "        evaluated_fitness = self._calculate_fitness(model, False)\n",
        "        print(\"  Aptitud lograda:\\t{}\\n\".format(evaluated_fitness))\n",
        "\n",
        "    def duplicate(self) -> RNAEnvironment:\n",
        "        \"\"\"\n",
        "        @return: New instance of the environment with identical parameters\n",
        "        \"\"\"\n",
        "        x = self.x_train.extend(self.x_val)\n",
        "        y = self.y_train.extend(self.y_val)\n",
        "        if hasattr(self, 'epochs'):\n",
        "            return RNAEnvironment(True, data_x=x, data_y=y, verbosity=self.verbosity, epochs=self.epochs, batch_size=self.batch_size)\n",
        "        else:\n",
        "            return RNAEnvironment(False, data_x=x, data_y=y, verbosity=self.verbosity)\n",
        "\n",
        "    def get_input_shape(self) -> (int,):\n",
        "        \"\"\"\"\"\"\n",
        "        return (self.cantX,)\n",
        "\n",
        "    def get_output_shape(self) -> (int,):\n",
        "        \"\"\"\"\"\"\n",
        "        return (self.cantY,)\n",
        "\n",
        "print(\"Clase RNAEnvironment definida\")"
      ],
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Clase RNAEnvironment definida\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "USZcQIWbIpEY"
      },
      "source": [
        "3) Ejecutar NeuroEvolution:"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "kHIt18d-AUTc",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "cellView": "form",
        "outputId": "6ee973b0-4b9e-480e-f1d0-3b89c7d900f7"
      },
      "source": [
        "#@title Ejecutar NeuroEvolution\n",
        "\n",
        "logging_level = logging.INFO\n",
        "config_file_path = confFileName # './codeepneat_xor_basic_example_config.cfg'\n",
        "backup_dir_path = './tfne_state_backups/'\n",
        "max_generations = ce_maximo_generaciones_procesar\n",
        "\n",
        "# Set logging, parse config\n",
        "logging.set_verbosity(logging_level)\n",
        "config = tfne.parse_configuration(config_file_path)\n",
        "\n",
        "# Initialize the environment and the specific NE algorithm\n",
        "environment = RNAEnvironment(weight_training=True, \n",
        "                             data_x=x_train,\n",
        "                             data_y=y_train,\n",
        "                             config=config, \n",
        "                             verbosity=logging_level)\n",
        "\n",
        "print(\">  Definiendo configuración:\")\n",
        "ne_algorithm = tfne.algorithms.CoDeepNEAT(config)\n",
        "\n",
        "# determina si termina por alcanzar la máxima o no\n",
        "if ce_finalizar_al_encontrar_max_fitness:\n",
        "  max_fitness = environment.return_MaxFitness()\n",
        "else:\n",
        "  max_fitness = None\n",
        "\n",
        "\n",
        "# Initialize evolution engine and supply config as well as initialized NE algorithm and evaluation environment.\n",
        "engine = tfne.EvolutionEngine(ne_algorithm=ne_algorithm,\n",
        "                              environment=environment,\n",
        "                              backup_dir_path=backup_dir_path,\n",
        "                              max_generations=max_generations,\n",
        "                              max_fitness=max_fitness)\n",
        "\n",
        "# Start training process, returning the best genome when training ends\n",
        "best_genome = engine.train()\n",
        "print(\"\\n> Mejor RNA generada por la evolución:\\n\")\n",
        "print(best_genome)\n",
        "print(\"\\n\")\n",
        "\n",
        "# Graba mejor genotipo y su modelo TF antes del re-entrenamiento\n",
        "best_genotype_ori_dir = './best_genome_genotype_ori/'\n",
        "best_model_ori_dir = './best_genome_model_ori/'\n",
        "best_genome.save_genotype(save_dir_path=best_genotype_ori_dir)\n",
        "best_genome.save_model(file_path=best_model_ori_dir)\n",
        "\n",
        "# Re-entrena mejor genotipo para ver si mejora\n",
        "print(\"\\n\\n\")\n",
        "environment.epochs = rna_cant_epocas_reentrenamiento_best_final\n",
        "print(\"> Re-entrenando mejor RNA por \"+str(environment.epochs)+\" épocas...\\n\")\n",
        "environment.eval_genome_fitness(best_genome)\n",
        "environment.replay_genome(best_genome)\n",
        "\n",
        "# Graba mejor genotipo y su modelo TF después del re-entrenamiento\n",
        "print(\"\\n\")\n",
        "best_genotype_new_dir = './best_genome_genotype/'\n",
        "best_model_new_dir = './best_genome_model/'\n",
        "best_genome.save_genotype(save_dir_path=best_genotype_new_dir)\n",
        "best_genome.save_model(file_path=best_model_new_dir)\n"
      ],
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "> Preparando el ambiente...\n",
            "Config value for 'EVALUATION/epochs': 30\n",
            "Config value for 'EVALUATION/batch_size': None\n",
            "Config value for 'EVALUATION/increase_epochs_every_n_genomes': 72\n",
            "Config value for 'EVALUATION/increase_epochs_add_epochs': 15\n",
            "Config value for 'FITNESS/calc_using_accuracy': S\n",
            "Config value for 'FITNESS/penalize_based_on_topology': True\n",
            " Definiendo parámetros para cálculo de aptitud: \n",
            "        calcFitness =  S\n",
            "        penalFitness =  True\n",
            "  Definiendo datos: de los  34 ejemplos de entrenamiento: \n",
            "                      se usan  30 ejemplos para entrenar \n",
            "                      y  4 ejemplos para validar.\n",
            "\n",
            "\n",
            ">  Definiendo configuración:\n",
            "Config value for 'POPULATION/bp_pop_size': 9\n",
            "Config value for 'POPULATION/mod_pop_size': 25\n",
            "Config value for 'POPULATION/genomes_per_bp': 8\n",
            "Config value for 'GENOME/dtype': float32\n",
            "Config value for 'GENOME/available_modules': ['DenseDropout']\n",
            "Config value for 'GENOME/available_optimizers': ['SGD', 'Adam']\n",
            "Config value for 'GENOME/output_layers': [{'class_name': 'Dense', 'config': {'units': 8, 'activation': 'softmax'}}]\n",
            "Config value for 'MODULE_SPECIATION/mod_spec_type': param-distance-dynamic\n",
            "Config value for 'MODULE_SPECIATION/mod_spec_species_count': 4\n",
            "Config value for 'MODULE_SPECIATION/mod_spec_distance': 0.3\n",
            "Config value for 'MODULE_SPECIATION/mod_spec_mod_elitism': 2\n",
            "Config value for 'MODULE_SPECIATION/mod_spec_min_offspring': 1\n",
            "Config value for 'MODULE_SPECIATION/mod_spec_reprod_thres': 0.5\n",
            "Config value for 'MODULE_SPECIATION/mod_spec_max_stagnation': 10\n",
            "Config value for 'MODULE_SPECIATION/mod_spec_species_elitism': 2\n",
            "Config value for 'MODULE_SPECIATION/mod_spec_rebase_repr': True\n",
            "Config value for 'MODULE_SPECIATION/mod_spec_reinit_extinct': False\n",
            "Config value for 'MODULE_EVOLUTION/mod_max_mutation': 0.3\n",
            "Config value for 'MODULE_EVOLUTION/mod_mutation_prob': 0.4\n",
            "Config value for 'MODULE_EVOLUTION/mod_crossover_prob': 0.6\n",
            "Config value for 'BP_SPECIATION/bp_spec_type': gene-overlap-dynamic\n",
            "Config value for 'BP_SPECIATION/bp_spec_species_count': 3\n",
            "Config value for 'BP_SPECIATION/bp_spec_distance': 0.3\n",
            "Config value for 'BP_SPECIATION/bp_spec_bp_elitism': 2\n",
            "Config value for 'BP_SPECIATION/bp_spec_min_offspring': 1\n",
            "Config value for 'BP_SPECIATION/bp_spec_reprod_thres': 0.5\n",
            "Config value for 'BP_SPECIATION/bp_spec_max_stagnation': 15\n",
            "Config value for 'BP_SPECIATION/bp_spec_species_elitism': 2\n",
            "Config value for 'BP_SPECIATION/bp_spec_rebase_repr': True\n",
            "Config value for 'BP_SPECIATION/bp_spec_reinit_extinct': True\n",
            "Config value for 'BP_EVOLUTION/bp_max_mutation': 0.3\n",
            "Config value for 'BP_EVOLUTION/bp_mutation_add_conn_prob': 0.2\n",
            "Config value for 'BP_EVOLUTION/bp_mutation_add_node_prob': 0.2\n",
            "Config value for 'BP_EVOLUTION/bp_mutation_rem_conn_prob': 0.05\n",
            "Config value for 'BP_EVOLUTION/bp_mutation_rem_node_prob': 0.05\n",
            "Config value for 'BP_EVOLUTION/bp_mutation_node_spec_prob': 0.3\n",
            "Config value for 'BP_EVOLUTION/bp_mutation_optimizer_prob': 0.1\n",
            "Config value for 'BP_EVOLUTION/bp_crossover_prob': 0.1\n",
            "Config value for 'MODULE_DENSEDROPOUT/merge_method': [{'class_name': 'Concatenate', 'config': {'axis': -1}}]\n",
            "Config value for 'MODULE_DENSEDROPOUT/units': {'min': 2, 'max': 100, 'step': 4, 'stddev': 6}\n",
            "Config value for 'MODULE_DENSEDROPOUT/activation': ['linear', 'relu', 'sigmoid', 'softmax', 'tanh']\n",
            "Config value for 'MODULE_DENSEDROPOUT/kernel_init': ['glorot_normal', 'he_normal']\n",
            "Config value for 'MODULE_DENSEDROPOUT/bias_init': ['zeros']\n",
            "Config value for 'MODULE_DENSEDROPOUT/dropout_flag': 0.5\n",
            "Config value for 'MODULE_DENSEDROPOUT/dropout_rate': {'min': 0.1, 'max': 0.4, 'step': 0.1, 'stddev': 0.1}\n",
            "Config value for 'OPTIMIZER_SGD/learning_rate': {'min': 0.1, 'max': 0.3, 'step': 0.05, 'stddev': 0.05}\n",
            "Config value for 'OPTIMIZER_SGD/momentum': {'min': 0.3, 'max': 0.7, 'step': 0.1, 'stddev': 0.1}\n",
            "Config value for 'OPTIMIZER_SGD/nesterov': [True, False]\n",
            "Config value for 'OPTIMIZER_ADAM/learning_rate': {'min': 0.0001, 'max': 0.1, 'step': 0.0001, 'stddev': 0.02}\n",
            "Config value for 'OPTIMIZER_ADAM/beta_1': {'min': 0.6, 'max': 1.5, 'step': 0.05, 'stddev': 0.2}\n",
            "Config value for 'OPTIMIZER_ADAM/beta_2': {'min': 0.8, 'max': 1.2, 'step': 0.001, 'stddev': 0.1}\n",
            "Config value for 'OPTIMIZER_ADAM/epsilon': {'min': 1e-08, 'max': 1e-06, 'step': 1e-08, 'stddev': 1e-07}\n",
            "Using Neuroevolution algorithm: CoDeepNEAT\n",
            "Using evaluation environment: RNAEnvironment\n",
            "Maximum number of generations to evolve the population: 15\n",
            "Maximum fitness value to evolve population up to: 210.0\n",
            "Creating TFNE generational Backups to directory: /content/tfne_state_backups/tfne_state_backup_2021-Sep-27_20-39-27/\n",
            "Initializing a new population of 9 blueprints and 25 modules...\n",
            "\n",
            "Evaluating 72 genomes in generation 0...\n",
            "[========================================] 72/72 Genomes | Genome ID 72 achieved fitness of 175.0\n",
            "\n",
            "\n",
            "\u001b[1m############################################################  Population Summary  ############################################################\n",
            "\n",
            "Generation:    0  ||  Best Genome Fitness:    200.0  ||  Avg Blueprint Fitness: 135.6713  ||  Avg Module Fitness: 116.9722\u001b[0m\n",
            "Best Genome: CoDeepNEAT Genome | ID:      7 | Fitness:  200.0 | Blueprint ID:      1 | Module Species: {1} | Optimizer:   adam | Origin Gen:    0\n",
            "\n",
            "\u001b[1mBlueprint Species       || Blueprint Species Avg Fitness       || Blueprint Species Size\u001b[0m\n",
            "     1                  || 135.6713                            ||        9\n",
            "Best BP of Species 1    || CoDeepNEAT Blueprint | ID:     #1 | Fitness:  175.0 | Nodes:    2 | Module Species: {1} | Optimizer: adam\n",
            "\n",
            "\u001b[1mModule Species          || Module Species Avg Fitness          || Module Species Size\u001b[0m\n",
            "     1                  || 116.9722                            ||       25\n",
            "Best Mod of Species 1   || CoDeepNEAT DENSE Module | ID:     #7 | Fitness:  175.0 | Units:   60 | Activ:   tanh | Dropout: None\n",
            "\n",
            "\u001b[1m##############################################################################################################################################\u001b[0m\n",
            "\n",
            "Backed up generation 0 of the CoDeepNEAT evolutionary run to file: /content/tfne_state_backups/tfne_state_backup_2021-Sep-27_20-39-27/tfne_state_backup_gen_0.json\n",
            "\n",
            "Evaluating 72 genomes in generation 1...\n",
            "[========================================] 72/72 Genomes | Genome ID 144 achieved fitness of 146.6667\n",
            "\n",
            "\n",
            "\u001b[1m############################################################  Population Summary  ############################################################\n",
            "\n",
            "Generation:    1  ||  Best Genome Fitness:   209.91  ||  Avg Blueprint Fitness: 167.5875  ||  Avg Module Fitness: 107.5764\u001b[0m\n",
            "Best Genome: CoDeepNEAT Genome | ID:    139 | Fitness: 209.91 | Blueprint ID:     16 | Module Species: {1} | Optimizer:    sgd | Origin Gen:    1\n",
            "\n",
            "\u001b[1mBlueprint Species       || Blueprint Species Avg Fitness       || Blueprint Species Size\u001b[0m\n",
            "     1                  || 167.5875                            ||        9\n",
            "Best BP of Species 1    || CoDeepNEAT Blueprint | ID:    #16 | Fitness: 175.8221 | Nodes:    2 | Module Species: {1} | Optimizer: sgd\n",
            "\n",
            "\u001b[1mModule Species          || Module Species Avg Fitness          || Module Species Size\u001b[0m\n",
            "     1                  || 168.0881                            ||       16\n",
            "Best Mod of Species 1   || CoDeepNEAT DENSE Module | ID:    #26 | Fitness: 180.8183 | Units:   24 | Activ:   tanh | Dropout:  0.2\n",
            "     2                  ||        0                            ||        2\n",
            "Best Mod of Species 2   || CoDeepNEAT DENSE Module | ID:    #33 | Fitness:      0 | Units:   76 | Activ:   tanh | Dropout: None\n",
            "     3                  ||        0                            ||        4\n",
            "Best Mod of Species 3   || CoDeepNEAT DENSE Module | ID:    #34 | Fitness:      0 | Units:   28 | Activ: linear | Dropout:  0.1\n",
            "     4                  ||        0                            ||        3\n",
            "Best Mod of Species 4   || CoDeepNEAT DENSE Module | ID:    #36 | Fitness:      0 | Units:   48 | Activ:   relu | Dropout: None\n",
            "\n",
            "\u001b[1m##############################################################################################################################################\u001b[0m\n",
            "\n",
            "Backed up generation 1 of the CoDeepNEAT evolutionary run to file: /content/tfne_state_backups/tfne_state_backup_2021-Sep-27_20-39-27/tfne_state_backup_gen_1.json\n",
            "\n",
            "Evaluating 72 genomes in generation 2...\n",
            "[========================================] 72/72 Genomes | Genome ID 216 achieved fitness of 175.0\n",
            "\n",
            "\n",
            "\u001b[1m############################################################  Population Summary  ############################################################\n",
            "\n",
            "Generation:    2  ||  Best Genome Fitness:   209.91  ||  Avg Blueprint Fitness: 169.9027  ||  Avg Module Fitness: 128.5112\u001b[0m\n",
            "Best Genome: CoDeepNEAT Genome | ID:    139 | Fitness: 209.91 | Blueprint ID:     16 | Module Species: {1} | Optimizer:    sgd | Origin Gen:    1\n",
            "\n",
            "\u001b[1mBlueprint Species       || Blueprint Species Avg Fitness       || Blueprint Species Size\u001b[0m\n",
            "     1                  || 169.9027                            ||        9\n",
            "Best BP of Species 1    || CoDeepNEAT Blueprint | ID:    #17 | Fitness: 181.7246 | Nodes:    3 | Module Species: {1} | Optimizer: sgd\n",
            "\n",
            "\u001b[1mModule Species          || Module Species Avg Fitness          || Module Species Size\u001b[0m\n",
            "     1                  ||  170.671                            ||       10\n",
            "Best Mod of Species 1   || CoDeepNEAT DENSE Module | ID:    #51 | Fitness: 181.9393 | Units:   36 | Activ:   tanh | Dropout: None\n",
            "     2                  ||        0                            ||        6\n",
            "Best Mod of Species 2   || CoDeepNEAT DENSE Module | ID:    #33 | Fitness:      0 | Units:   76 | Activ:   tanh | Dropout: None\n",
            "     3                  || 164.1667                            ||        5\n",
            "Best Mod of Species 3   || CoDeepNEAT DENSE Module | ID:    #42 | Fitness:  175.0 | Units:   24 | Activ: linear | Dropout: None\n",
            "     4                  || 171.3094                            ||        4\n",
            "Best Mod of Species 4   || CoDeepNEAT DENSE Module | ID:    #36 | Fitness: 192.295 | Units:   48 | Activ:   relu | Dropout: None\n",
            "\n",
            "\u001b[1m##############################################################################################################################################\u001b[0m\n",
            "\n",
            "Backed up generation 2 of the CoDeepNEAT evolutionary run to file: /content/tfne_state_backups/tfne_state_backup_2021-Sep-27_20-39-27/tfne_state_backup_gen_2.json\n",
            "\n",
            "Evaluating 72 genomes in generation 3...\n",
            "[========================================] 72/72 Genomes | Genome ID 288 achieved fitness of 175.0\n",
            "\n",
            "\n",
            "\u001b[1m############################################################  Population Summary  ############################################################\n",
            "\n",
            "Generation:    3  ||  Best Genome Fitness: 209.93667  ||  Avg Blueprint Fitness: 173.5907  ||  Avg Module Fitness: 172.3683\u001b[0m\n",
            "Best Genome: CoDeepNEAT Genome | ID:    262 | Fitness: 209.93667 | Blueprint ID:     27 | Module Species: {1} | Optimizer:    sgd | Origin Gen:    3\n",
            "\n",
            "\u001b[1mBlueprint Species       || Blueprint Species Avg Fitness       || Blueprint Species Size\u001b[0m\n",
            "     1                  || 173.5907                            ||        9\n",
            "Best BP of Species 1    || CoDeepNEAT Blueprint | ID:    #30 | Fitness: 180.5425 | Nodes:    3 | Module Species: {4} | Optimizer: adam\n",
            "\n",
            "\u001b[1mModule Species          || Module Species Avg Fitness          || Module Species Size\u001b[0m\n",
            "     1                  || 172.8444                            ||        9\n",
            "Best Mod of Species 1   || CoDeepNEAT DENSE Module | ID:    #67 | Fitness: 192.3675 | Units:   32 | Activ:   tanh | Dropout: None\n",
            "     2                  ||    170.0                            ||        3\n",
            "Best Mod of Species 2   || CoDeepNEAT DENSE Module | ID:    #53 | Fitness:  175.0 | Units:   36 | Activ:   tanh | Dropout: 0.30000000000000004\n",
            "     3                  || 172.9167                            ||        6\n",
            "Best Mod of Species 3   || CoDeepNEAT DENSE Module | ID:    #58 | Fitness:  175.0 | Units:   40 | Activ:   tanh | Dropout: None\n",
            "     4                  || 172.3012                            ||        7\n",
            "Best Mod of Species 4   || CoDeepNEAT DENSE Module | ID:    #76 | Fitness: 183.6542 | Units:   40 | Activ:   relu | Dropout: None\n",
            "\n",
            "\u001b[1m##############################################################################################################################################\u001b[0m\n",
            "\n",
            "Backed up generation 3 of the CoDeepNEAT evolutionary run to file: /content/tfne_state_backups/tfne_state_backup_2021-Sep-27_20-39-27/tfne_state_backup_gen_3.json\n",
            "\n",
            "Evaluating 72 genomes in generation 4...\n",
            "[========================================] 72/72 Genomes | Genome ID 360 achieved fitness of 150.0\n",
            "\n",
            "\n",
            "\u001b[1m############################################################  Population Summary  ############################################################\n",
            "\n",
            "Generation:    4  ||  Best Genome Fitness: 209.93667  ||  Avg Blueprint Fitness: 173.2748  ||  Avg Module Fitness: 165.2517\u001b[0m\n",
            "Best Genome: CoDeepNEAT Genome | ID:    262 | Fitness: 209.93667 | Blueprint ID:     27 | Module Species: {1} | Optimizer:    sgd | Origin Gen:    3\n",
            "\n",
            "\u001b[1mBlueprint Species       || Blueprint Species Avg Fitness       || Blueprint Species Size\u001b[0m\n",
            "     1                  || 173.2748                            ||        9\n",
            "Best BP of Species 1    || CoDeepNEAT Blueprint | ID:    #35 | Fitness: 183.7342 | Nodes:    2 | Module Species: {1} | Optimizer: sgd\n",
            "\n",
            "\u001b[1mModule Species          || Module Species Avg Fitness          || Module Species Size\u001b[0m\n",
            "     1                  || 175.4694                            ||        4\n",
            "Best Mod of Species 1   || CoDeepNEAT DENSE Module | ID:    #63 | Fitness: 179.991 | Units:   32 | Activ:   tanh | Dropout: None\n",
            "     2                  || 174.1755                            ||        6\n",
            "Best Mod of Species 2   || CoDeepNEAT DENSE Module | ID:    #77 | Fitness: 189.63 | Units:   60 | Activ:   tanh | Dropout:  0.2\n",
            "     3                  || 145.8333                            ||        8\n",
            "Best Mod of Species 3   || CoDeepNEAT DENSE Module | ID:    #58 | Fitness:  175.0 | Units:   40 | Activ:   tanh | Dropout: None\n",
            "     4                  || 173.9566                            ||        7\n",
            "Best Mod of Species 4   || CoDeepNEAT DENSE Module | ID:    #76 | Fitness: 182.2211 | Units:   40 | Activ:   relu | Dropout: None\n",
            "\n",
            "\u001b[1m##############################################################################################################################################\u001b[0m\n",
            "\n",
            "Backed up generation 4 of the CoDeepNEAT evolutionary run to file: /content/tfne_state_backups/tfne_state_backup_2021-Sep-27_20-39-27/tfne_state_backup_gen_4.json\n",
            "\n",
            "Evaluating 72 genomes in generation 5...\n",
            "[========================================] 72/72 Genomes | Genome ID 432 achieved fitness of 175.0\n",
            "\n",
            "\n",
            "\u001b[1m############################################################  Population Summary  ############################################################\n",
            "\n",
            "Generation:    5  ||  Best Genome Fitness: 209.93667  ||  Avg Blueprint Fitness: 154.5133  ||  Avg Module Fitness: 132.7506\u001b[0m\n",
            "Best Genome: CoDeepNEAT Genome | ID:    262 | Fitness: 209.93667 | Blueprint ID:     27 | Module Species: {1} | Optimizer:    sgd | Origin Gen:    3\n",
            "\n",
            "\u001b[1mBlueprint Species       || Blueprint Species Avg Fitness       || Blueprint Species Size\u001b[0m\n",
            "     1                  || 175.8841                            ||        4\n",
            "Best BP of Species 1    || CoDeepNEAT Blueprint | ID:    #36 | Fitness: 183.5442 | Nodes:    4 | Module Species: {1, 2, 4} | Optimizer: sgd\n",
            "     2                  || 137.4167                            ||        5\n",
            "Best BP of Species 2    || CoDeepNEAT Blueprint | ID:    #42 | Fitness:  175.0 | Nodes:    3 | Module Species: {1, 2} | Optimizer: sgd\n",
            "\n",
            "\u001b[1mModule Species          || Module Species Avg Fitness          || Module Species Size\u001b[0m\n",
            "     1                  || 149.2414                            ||        3\n",
            "Best Mod of Species 1   || CoDeepNEAT DENSE Module | ID:    #67 | Fitness: 157.5792 | Units:   32 | Activ:   tanh | Dropout: None\n",
            "     2                  || 146.2064                            ||        5\n",
            "Best Mod of Species 2   || CoDeepNEAT DENSE Module | ID:    #33 | Fitness:  157.5 | Units:   76 | Activ:   tanh | Dropout: None\n",
            "     3                  || 133.3333                            ||        9\n",
            "Best Mod of Species 3   || CoDeepNEAT DENSE Module | ID:    #58 | Fitness:  175.0 | Units:   40 | Activ:   tanh | Dropout: None\n",
            "     4                  || 117.5013                            ||        8\n",
            "Best Mod of Species 4   || CoDeepNEAT DENSE Module | ID:    #96 | Fitness: 192.0817 | Units:   44 | Activ:   relu | Dropout: None\n",
            "\n",
            "\u001b[1m##############################################################################################################################################\u001b[0m\n",
            "\n",
            "Backed up generation 5 of the CoDeepNEAT evolutionary run to file: /content/tfne_state_backups/tfne_state_backup_2021-Sep-27_20-39-27/tfne_state_backup_gen_5.json\n",
            "\n",
            "Evaluating 72 genomes in generation 6...\n",
            "[========================================] 72/72 Genomes | Genome ID 504 achieved fitness of 175.0\n",
            "\n",
            "\n",
            "\u001b[1m############################################################  Population Summary  ############################################################\n",
            "\n",
            "Generation:    6  ||  Best Genome Fitness: 209.93667  ||  Avg Blueprint Fitness: 176.8351  ||  Avg Module Fitness: 161.9024\u001b[0m\n",
            "Best Genome: CoDeepNEAT Genome | ID:    262 | Fitness: 209.93667 | Blueprint ID:     27 | Module Species: {1} | Optimizer:    sgd | Origin Gen:    3\n",
            "\n",
            "\u001b[1mBlueprint Species       || Blueprint Species Avg Fitness       || Blueprint Species Size\u001b[0m\n",
            "     1                  || 172.4494                            ||        2\n",
            "Best BP of Species 1    || CoDeepNEAT Blueprint | ID:    #36 | Fitness: 176.1488 | Nodes:    4 | Module Species: {1, 2, 4} | Optimizer: sgd\n",
            "     2                  || 177.8916                            ||        6\n",
            "Best BP of Species 2    || CoDeepNEAT Blueprint | ID:    #46 | Fitness: 183.6067 | Nodes:    3 | Module Species: {1} | Optimizer: sgd\n",
            "     3                  || 179.2671                            ||        1\n",
            "Best BP of Species 3    || CoDeepNEAT Blueprint | ID:    #49 | Fitness: 179.2671 | Nodes:    4 | Module Species: {1, 2, 4} | Optimizer: sgd\n",
            "\n",
            "\u001b[1mModule Species          || Module Species Avg Fitness          || Module Species Size\u001b[0m\n",
            "     1                  || 176.5826                            ||        3\n",
            "Best Mod of Species 1   || CoDeepNEAT DENSE Module | ID:    #69 | Fitness: 178.4196 | Units:   32 | Activ:   tanh | Dropout: None\n",
            "     2                  || 176.6063                            ||        7\n",
            "Best Mod of Species 2   || CoDeepNEAT DENSE Module | ID:   #115 | Fitness: 184.28 | Units:   68 | Activ:   tanh | Dropout: None\n",
            "     3                  || 172.8416                            ||        8\n",
            "Best Mod of Species 3   || CoDeepNEAT DENSE Module | ID:   #116 | Fitness: 192.135 | Units:   32 | Activ:   tanh | Dropout: None\n",
            "     4                  || 179.7672                            ||        5\n",
            "Best Mod of Species 4   || CoDeepNEAT DENSE Module | ID:   #105 | Fitness: 183.5342 | Units:   44 | Activ:   relu | Dropout: None\n",
            "     5                  ||        0                            ||        1\n",
            "Best Mod of Species 5   || CoDeepNEAT DENSE Module | ID:   #119 | Fitness:      0 | Units:   32 | Activ: softmax | Dropout: None\n",
            "     6                  ||        0                            ||        1\n",
            "Best Mod of Species 6   || CoDeepNEAT DENSE Module | ID:   #120 | Fitness:      0 | Units:   32 | Activ:   relu | Dropout: None\n",
            "\n",
            "\u001b[1m##############################################################################################################################################\u001b[0m\n",
            "\n",
            "Backed up generation 6 of the CoDeepNEAT evolutionary run to file: /content/tfne_state_backups/tfne_state_backup_2021-Sep-27_20-39-27/tfne_state_backup_gen_6.json\n",
            "\n",
            "Evaluating 72 genomes in generation 7...\n",
            "[========================================] 72/72 Genomes | Genome ID 576 achieved fitness of 143.3333\n",
            "\n",
            "\n",
            "\u001b[1m############################################################  Population Summary  ############################################################\n",
            "\n",
            "Generation:    7  ||  Best Genome Fitness: 209.93667  ||  Avg Blueprint Fitness: 173.4284  ||  Avg Module Fitness: 162.3004\u001b[0m\n",
            "Best Genome: CoDeepNEAT Genome | ID:    262 | Fitness: 209.93667 | Blueprint ID:     27 | Module Species: {1} | Optimizer:    sgd | Origin Gen:    3\n",
            "\n",
            "\u001b[1mBlueprint Species       || Blueprint Species Avg Fitness       || Blueprint Species Size\u001b[0m\n",
            "     1                  || 174.3401                            ||        3\n",
            "Best BP of Species 1    || CoDeepNEAT Blueprint | ID:    #36 | Fitness: 179.2704 | Nodes:    4 | Module Species: {1, 2, 4} | Optimizer: sgd\n",
            "     2                  || 177.7541                            ||        2\n",
            "Best BP of Species 2    || CoDeepNEAT Blueprint | ID:    #46 | Fitness: 180.5083 | Nodes:    3 | Module Species: {1} | Optimizer: sgd\n",
            "     3                  || 170.5817                            ||        4\n",
            "Best BP of Species 3    || CoDeepNEAT Blueprint | ID:    #49 | Fitness: 183.5375 | Nodes:    4 | Module Species: {1, 2, 4} | Optimizer: sgd\n",
            "\n",
            "\u001b[1mModule Species          || Module Species Avg Fitness          || Module Species Size\u001b[0m\n",
            "     1                  ||  173.487                            ||        6\n",
            "Best Mod of Species 1   || CoDeepNEAT DENSE Module | ID:   #126 | Fitness: 177.8003 | Units:   32 | Activ:   tanh | Dropout: None\n",
            "     2                  || 172.2179                            ||        5\n",
            "Best Mod of Species 2   || CoDeepNEAT DENSE Module | ID:   #128 | Fitness: 177.7772 | Units:   72 | Activ:   tanh | Dropout: None\n",
            "     3                  || 136.3082                            ||        4\n",
            "Best Mod of Species 3   || CoDeepNEAT DENSE Module | ID:   #116 | Fitness: 192.135 | Units:   32 | Activ:   tanh | Dropout: None\n",
            "     4                  || 172.2666                            ||        6\n",
            "Best Mod of Species 4   || CoDeepNEAT DENSE Module | ID:    #96 | Fitness: 181.7558 | Units:   44 | Activ:   relu | Dropout: None\n",
            "     5                  || 142.7778                            ||        2\n",
            "Best Mod of Species 5   || CoDeepNEAT DENSE Module | ID:   #119 | Fitness:  147.5 | Units:   32 | Activ: softmax | Dropout: None\n",
            "     6                  || 145.5556                            ||        2\n",
            "Best Mod of Species 6   || CoDeepNEAT DENSE Module | ID:   #122 | Fitness: 155.8334 | Units:   32 | Activ:   relu | Dropout: None\n",
            "\n",
            "\u001b[1m##############################################################################################################################################\u001b[0m\n",
            "\n",
            "Backed up generation 7 of the CoDeepNEAT evolutionary run to file: /content/tfne_state_backups/tfne_state_backup_2021-Sep-27_20-39-27/tfne_state_backup_gen_7.json\n",
            "\n",
            "Evaluating 80 genomes in generation 8...\n",
            "[========================================] 80/80 Genomes | Genome ID 656 achieved fitness of 175.0\n",
            "\n",
            "\n",
            "\u001b[1m############################################################  Population Summary  ############################################################\n",
            "\n",
            "Generation:    8  ||  Best Genome Fitness: 209.93667  ||  Avg Blueprint Fitness:  174.748  ||  Avg Module Fitness: 158.7417\u001b[0m\n",
            "Best Genome: CoDeepNEAT Genome | ID:    262 | Fitness: 209.93667 | Blueprint ID:     27 | Module Species: {1} | Optimizer:    sgd | Origin Gen:    3\n",
            "\n",
            "\u001b[1mBlueprint Species       || Blueprint Species Avg Fitness       || Blueprint Species Size\u001b[0m\n",
            "     1                  ||  173.748                            ||        4\n",
            "Best BP of Species 1    || CoDeepNEAT Blueprint | ID:    #35 | Fitness: 176.2421 | Nodes:    2 | Module Species: {1} | Optimizer: sgd\n",
            "     2                  || 174.0333                            ||        2\n",
            "Best BP of Species 2    || CoDeepNEAT Blueprint | ID:    #47 | Fitness:  175.0 | Nodes:    4 | Module Species: {1, 2, 4} | Optimizer: sgd\n",
            "     3                  || 176.1054                            ||        4\n",
            "Best BP of Species 3    || CoDeepNEAT Blueprint | ID:    #54 | Fitness: 180.34 | Nodes:    4 | Module Species: {1, 2, 4} | Optimizer: sgd\n",
            "\n",
            "\u001b[1mModule Species          || Module Species Avg Fitness          || Module Species Size\u001b[0m\n",
            "     1                  || 175.3973                            ||        5\n",
            "Best Mod of Species 1   || CoDeepNEAT DENSE Module | ID:    #69 | Fitness: 179.542 | Units:   32 | Activ:   tanh | Dropout: None\n",
            "     2                  || 175.7502                            ||        4\n",
            "Best Mod of Species 2   || CoDeepNEAT DENSE Module | ID:   #139 | Fitness: 176.1075 | Units:   72 | Activ:   tanh | Dropout: None\n",
            "     3                  || 136.3082                            ||        4\n",
            "Best Mod of Species 3   || CoDeepNEAT DENSE Module | ID:   #116 | Fitness: 192.135 | Units:   32 | Activ:   tanh | Dropout: None\n",
            "     4                  || 175.4424                            ||        5\n",
            "Best Mod of Species 4   || CoDeepNEAT DENSE Module | ID:   #137 | Fitness: 179.8486 | Units:   32 | Activ:   relu | Dropout: None\n",
            "     5                  ||   168.75                            ||        4\n",
            "Best Mod of Species 5   || CoDeepNEAT DENSE Module | ID:   #119 | Fitness:  175.0 | Units:   32 | Activ: softmax | Dropout: None\n",
            "     6                  ||  97.0371                            ||        3\n",
            "Best Mod of Species 6   || CoDeepNEAT DENSE Module | ID:   #122 | Fitness: 155.8334 | Units:   32 | Activ:   relu | Dropout: None\n",
            "\n",
            "\u001b[1m##############################################################################################################################################\u001b[0m\n",
            "\n",
            "Backed up generation 8 of the CoDeepNEAT evolutionary run to file: /content/tfne_state_backups/tfne_state_backup_2021-Sep-27_20-39-27/tfne_state_backup_gen_8.json\n",
            "\n",
            "Evaluating 72 genomes in generation 9...\n",
            "[========================================] 72/72 Genomes | Genome ID 728 achieved fitness of 150.0\n",
            "\n",
            "\n",
            "\u001b[1m############################################################  Population Summary  ############################################################\n",
            "\n",
            "Generation:    9  ||  Best Genome Fitness: 209.93667  ||  Avg Blueprint Fitness: 175.0773  ||  Avg Module Fitness: 165.1504\u001b[0m\n",
            "Best Genome: CoDeepNEAT Genome | ID:    262 | Fitness: 209.93667 | Blueprint ID:     27 | Module Species: {1} | Optimizer:    sgd | Origin Gen:    3\n",
            "\n",
            "\u001b[1mBlueprint Species       || Blueprint Species Avg Fitness       || Blueprint Species Size\u001b[0m\n",
            "     1                  ||    175.0                            ||        2\n",
            "Best BP of Species 1    || CoDeepNEAT Blueprint | ID:    #35 | Fitness:  175.0 | Nodes:    2 | Module Species: {1} | Optimizer: sgd\n",
            "     2                  || 167.5076                            ||        3\n",
            "Best BP of Species 2    || CoDeepNEAT Blueprint | ID:    #47 | Fitness: 179.1896 | Nodes:    4 | Module Species: {1, 2, 4} | Optimizer: sgd\n",
            "     3                  || 180.7931                            ||        4\n",
            "Best BP of Species 3    || CoDeepNEAT Blueprint | ID:    #54 | Fitness: 187.6775 | Nodes:    4 | Module Species: {1, 2, 4} | Optimizer: sgd\n",
            "\n",
            "\u001b[1mModule Species          || Module Species Avg Fitness          || Module Species Size\u001b[0m\n",
            "     1                  || 174.8702                            ||        5\n",
            "Best Mod of Species 1   || CoDeepNEAT DENSE Module | ID:   #148 | Fitness: 179.0206 | Units:   32 | Activ:   tanh | Dropout: None\n",
            "     2                  || 182.8437                            ||        5\n",
            "Best Mod of Species 2   || CoDeepNEAT DENSE Module | ID:   #154 | Fitness: 186.2778 | Units:   72 | Activ:   tanh | Dropout: None\n",
            "     3                  || 179.2858                            ||        4\n",
            "Best Mod of Species 3   || CoDeepNEAT DENSE Module | ID:   #102 | Fitness: 208.81 | Units:   32 | Activ:   tanh | Dropout: None\n",
            "     4                  || 180.9698                            ||        4\n",
            "Best Mod of Species 4   || CoDeepNEAT DENSE Module | ID:   #151 | Fitness: 187.9459 | Units:   40 | Activ:   relu | Dropout: None\n",
            "     5                  || 152.0139                            ||        4\n",
            "Best Mod of Species 5   || CoDeepNEAT DENSE Module | ID:   #135 | Fitness:  165.0 | Units:   32 | Activ: softmax | Dropout: None\n",
            "     6                  ||  97.0371                            ||        3\n",
            "Best Mod of Species 6   || CoDeepNEAT DENSE Module | ID:   #122 | Fitness: 155.8334 | Units:   32 | Activ:   relu | Dropout: None\n",
            "\n",
            "\u001b[1m##############################################################################################################################################\u001b[0m\n",
            "\n",
            "Backed up generation 9 of the CoDeepNEAT evolutionary run to file: /content/tfne_state_backups/tfne_state_backup_2021-Sep-27_20-39-27/tfne_state_backup_gen_9.json\n",
            "\n",
            "Evaluating 72 genomes in generation 10...\n",
            "[========================================] 72/72 Genomes | Genome ID 800 achieved fitness of 150.0\n",
            "\n",
            "\n",
            "\u001b[1m############################################################  Population Summary  ############################################################\n",
            "\n",
            "Generation:   10  ||  Best Genome Fitness: 209.93667  ||  Avg Blueprint Fitness:  166.604  ||  Avg Module Fitness: 149.1972\u001b[0m\n",
            "Best Genome: CoDeepNEAT Genome | ID:    262 | Fitness: 209.93667 | Blueprint ID:     27 | Module Species: {1} | Optimizer:    sgd | Origin Gen:    3\n",
            "\n",
            "\u001b[1mBlueprint Species       || Blueprint Species Avg Fitness       || Blueprint Species Size\u001b[0m\n",
            "     1                  || 149.2361                            ||        3\n",
            "Best BP of Species 1    || CoDeepNEAT Blueprint | ID:    #35 | Fitness: 171.875 | Nodes:    2 | Module Species: {1} | Optimizer: sgd\n",
            "     2                  || 174.5656                            ||        2\n",
            "Best BP of Species 2    || CoDeepNEAT Blueprint | ID:    #47 | Fitness: 179.1896 | Nodes:    4 | Module Species: {1, 2, 4} | Optimizer: sgd\n",
            "     3                  || 175.6492                            ||        4\n",
            "Best BP of Species 3    || CoDeepNEAT Blueprint | ID:    #58 | Fitness: 191.6517 | Nodes:    5 | Module Species: {1, 2, 4} | Optimizer: sgd\n",
            "\n",
            "\u001b[1mModule Species          || Module Species Avg Fitness          || Module Species Size\u001b[0m\n",
            "     1                  || 172.5939                            ||        4\n",
            "Best Mod of Species 1   || CoDeepNEAT DENSE Module | ID:   #148 | Fitness: 175.3184 | Units:   32 | Activ:   tanh | Dropout: None\n",
            "     2                  || 176.7723                            ||        5\n",
            "Best Mod of Species 2   || CoDeepNEAT DENSE Module | ID:    #33 | Fitness: 180.5506 | Units:   76 | Activ:   tanh | Dropout: None\n",
            "     3                  || 108.4287                            ||        5\n",
            "Best Mod of Species 3   || CoDeepNEAT DENSE Module | ID:   #102 | Fitness: 208.81 | Units:   32 | Activ:   tanh | Dropout: None\n",
            "     4                  || 182.0216                            ||        4\n",
            "Best Mod of Species 4   || CoDeepNEAT DENSE Module | ID:   #160 | Fitness: 188.422 | Units:   40 | Activ:   relu | Dropout: None\n",
            "     5                  ||  148.588                            ||        4\n",
            "Best Mod of Species 5   || CoDeepNEAT DENSE Module | ID:   #135 | Fitness: 152.7778 | Units:   32 | Activ: softmax | Dropout: None\n",
            "     6                  ||  97.0371                            ||        3\n",
            "Best Mod of Species 6   || CoDeepNEAT DENSE Module | ID:   #122 | Fitness: 155.8334 | Units:   32 | Activ:   relu | Dropout: None\n",
            "\n",
            "\u001b[1m##############################################################################################################################################\u001b[0m\n",
            "\n",
            "Backed up generation 10 of the CoDeepNEAT evolutionary run to file: /content/tfne_state_backups/tfne_state_backup_2021-Sep-27_20-39-27/tfne_state_backup_gen_10.json\n",
            "\n",
            "Evaluating 72 genomes in generation 11...\n",
            "[========================================] 72/72 Genomes | Genome ID 872 achieved fitness of 175.0\n",
            "\n",
            "\n",
            "\u001b[1m############################################################  Population Summary  ############################################################\n",
            "\n",
            "Generation:   11  ||  Best Genome Fitness: 209.93667  ||  Avg Blueprint Fitness: 162.8698  ||  Avg Module Fitness: 149.1574\u001b[0m\n",
            "Best Genome: CoDeepNEAT Genome | ID:    262 | Fitness: 209.93667 | Blueprint ID:     27 | Module Species: {1} | Optimizer:    sgd | Origin Gen:    3\n",
            "\n",
            "\u001b[1mBlueprint Species       || Blueprint Species Avg Fitness       || Blueprint Species Size\u001b[0m\n",
            "     1                  || 163.5417                            ||        3\n",
            "Best BP of Species 1    || CoDeepNEAT Blueprint | ID:    #35 | Fitness: 168.75 | Nodes:    2 | Module Species: {1} | Optimizer: sgd\n",
            "     2                  || 142.7638                            ||        3\n",
            "Best BP of Species 2    || CoDeepNEAT Blueprint | ID:    #47 | Fitness: 179.1829 | Nodes:    4 | Module Species: {1, 2, 4} | Optimizer: sgd\n",
            "     3                  ||  182.304                            ||        3\n",
            "Best BP of Species 3    || CoDeepNEAT Blueprint | ID:    #58 | Fitness: 187.4721 | Nodes:    5 | Module Species: {1, 2, 4} | Optimizer: sgd\n",
            "\n",
            "\u001b[1mModule Species          || Module Species Avg Fitness          || Module Species Size\u001b[0m\n",
            "     1                  || 164.2417                            ||        5\n",
            "Best Mod of Species 1   || CoDeepNEAT DENSE Module | ID:   #168 | Fitness: 175.0112 | Units:   32 | Activ:   tanh | Dropout: None\n",
            "     2                  || 180.6852                            ||        4\n",
            "Best Mod of Species 2   || CoDeepNEAT DENSE Module | ID:    #33 | Fitness: 184.4962 | Units:   76 | Activ:   tanh | Dropout: None\n",
            "     3                  || 135.4196                            ||        4\n",
            "Best Mod of Species 3   || CoDeepNEAT DENSE Module | ID:   #145 | Fitness: 191.6783 | Units:   36 | Activ:   tanh | Dropout: None\n",
            "     4                  || 181.1798                            ||        5\n",
            "Best Mod of Species 4   || CoDeepNEAT DENSE Module | ID:    #96 | Fitness: 191.7317 | Units:   44 | Activ:   relu | Dropout: None\n",
            "     5                  || 111.5741                            ||        4\n",
            "Best Mod of Species 5   || CoDeepNEAT DENSE Module | ID:   #135 | Fitness: 152.7778 | Units:   32 | Activ: softmax | Dropout: None\n",
            "     6                  ||  97.0371                            ||        3\n",
            "Best Mod of Species 6   || CoDeepNEAT DENSE Module | ID:   #122 | Fitness: 155.8334 | Units:   32 | Activ:   relu | Dropout: None\n",
            "\n",
            "\u001b[1m##############################################################################################################################################\u001b[0m\n",
            "\n",
            "Backed up generation 11 of the CoDeepNEAT evolutionary run to file: /content/tfne_state_backups/tfne_state_backup_2021-Sep-27_20-39-27/tfne_state_backup_gen_11.json\n",
            "\n",
            "Evaluating 72 genomes in generation 12...\n",
            "[========================================] 72/72 Genomes | Genome ID 944 achieved fitness of 175.0\n",
            "\n",
            "\n",
            "\u001b[1m############################################################  Population Summary  ############################################################\n",
            "\n",
            "Generation:   12  ||  Best Genome Fitness: 209.93667  ||  Avg Blueprint Fitness: 158.5932  ||  Avg Module Fitness: 155.8638\u001b[0m\n",
            "Best Genome: CoDeepNEAT Genome | ID:    262 | Fitness: 209.93667 | Blueprint ID:     27 | Module Species: {1} | Optimizer:    sgd | Origin Gen:    3\n",
            "\n",
            "\u001b[1mBlueprint Species       || Blueprint Species Avg Fitness       || Blueprint Species Size\u001b[0m\n",
            "     1                  || 113.5417                            ||        3\n",
            "Best BP of Species 1    || CoDeepNEAT Blueprint | ID:    #35 | Fitness:  175.0 | Nodes:    2 | Module Species: {1} | Optimizer: sgd\n",
            "     2                  || 177.6873                            ||        2\n",
            "Best BP of Species 2    || CoDeepNEAT Blueprint | ID:    #46 | Fitness: 179.3167 | Nodes:    3 | Module Species: {1} | Optimizer: sgd\n",
            "     3                  || 182.8349                            ||        4\n",
            "Best BP of Species 3    || CoDeepNEAT Blueprint | ID:    #66 | Fitness: 192.8896 | Nodes:    4 | Module Species: {1, 2} | Optimizer: sgd\n",
            "\n",
            "\u001b[1mModule Species          || Module Species Avg Fitness          || Module Species Size\u001b[0m\n",
            "     1                  || 160.7465                            ||        4\n",
            "Best Mod of Species 1   || CoDeepNEAT DENSE Module | ID:   #168 | Fitness: 184.0711 | Units:   32 | Activ:   tanh | Dropout: None\n",
            "     2                  || 183.4269                            ||        5\n",
            "Best Mod of Species 2   || CoDeepNEAT DENSE Module | ID:   #177 | Fitness: 197.4156 | Units:   76 | Activ:   tanh | Dropout: None\n",
            "     3                  || 177.7797                            ||        4\n",
            "Best Mod of Species 3   || CoDeepNEAT DENSE Module | ID:   #175 | Fitness: 186.1189 | Units:   36 | Activ:   tanh | Dropout: None\n",
            "     4                  || 177.5898                            ||        5\n",
            "Best Mod of Species 4   || CoDeepNEAT DENSE Module | ID:    #96 | Fitness: 188.3427 | Units:   44 | Activ:   relu | Dropout: None\n",
            "     5                  || 111.5741                            ||        4\n",
            "Best Mod of Species 5   || CoDeepNEAT DENSE Module | ID:   #135 | Fitness: 152.7778 | Units:   32 | Activ: softmax | Dropout: None\n",
            "     6                  ||  97.0371                            ||        3\n",
            "Best Mod of Species 6   || CoDeepNEAT DENSE Module | ID:   #122 | Fitness: 155.8334 | Units:   32 | Activ:   relu | Dropout: None\n",
            "\n",
            "\u001b[1m##############################################################################################################################################\u001b[0m\n",
            "\n",
            "Backed up generation 12 of the CoDeepNEAT evolutionary run to file: /content/tfne_state_backups/tfne_state_backup_2021-Sep-27_20-39-27/tfne_state_backup_gen_12.json\n",
            "\n",
            "Evaluating 72 genomes in generation 13...\n",
            "[========================================] 72/72 Genomes | Genome ID 1016 achieved fitness of 175.0\n",
            "\n",
            "\n",
            "\u001b[1m############################################################  Population Summary  ############################################################\n",
            "\n",
            "Generation:   13  ||  Best Genome Fitness: 209.93667  ||  Avg Blueprint Fitness: 171.0902  ||  Avg Module Fitness: 148.4545\u001b[0m\n",
            "Best Genome: CoDeepNEAT Genome | ID:    262 | Fitness: 209.93667 | Blueprint ID:     27 | Module Species: {1} | Optimizer:    sgd | Origin Gen:    3\n",
            "\n",
            "\u001b[1mBlueprint Species       || Blueprint Species Avg Fitness       || Blueprint Species Size\u001b[0m\n",
            "     1                  || 166.6667                            ||        3\n",
            "Best BP of Species 1    || CoDeepNEAT Blueprint | ID:    #63 | Fitness: 171.875 | Nodes:    2 | Module Species: {1} | Optimizer: sgd\n",
            "     2                  || 174.5623                            ||        2\n",
            "Best BP of Species 2    || CoDeepNEAT Blueprint | ID:    #47 | Fitness: 179.1829 | Nodes:    4 | Module Species: {1, 2, 4} | Optimizer: sgd\n",
            "     3                  || 172.6718                            ||        4\n",
            "Best BP of Species 3    || CoDeepNEAT Blueprint | ID:    #58 | Fitness: 179.1562 | Nodes:    5 | Module Species: {1, 2, 4} | Optimizer: sgd\n",
            "\n",
            "\u001b[1mModule Species          || Module Species Avg Fitness          || Module Species Size\u001b[0m\n",
            "     1                  ||  170.908                            ||        4\n",
            "Best Mod of Species 1   || CoDeepNEAT DENSE Module | ID:   #168 | Fitness: 175.988 | Units:   32 | Activ:   tanh | Dropout: None\n",
            "     2                  || 172.5051                            ||        5\n",
            "Best Mod of Species 2   || CoDeepNEAT DENSE Module | ID:   #191 | Fitness: 179.2087 | Units:   76 | Activ:   tanh | Dropout: None\n",
            "     3                  || 134.0297                            ||        4\n",
            "Best Mod of Species 3   || CoDeepNEAT DENSE Module | ID:   #175 | Fitness: 186.1189 | Units:   36 | Activ:   tanh | Dropout: None\n",
            "     4                  || 178.3357                            ||        5\n",
            "Best Mod of Species 4   || CoDeepNEAT DENSE Module | ID:   #181 | Fitness: 191.6783 | Units:   44 | Activ:   relu | Dropout: None\n",
            "     5                  || 111.5741                            ||        4\n",
            "Best Mod of Species 5   || CoDeepNEAT DENSE Module | ID:   #135 | Fitness: 152.7778 | Units:   32 | Activ: softmax | Dropout: None\n",
            "     6                  ||  97.0371                            ||        3\n",
            "Best Mod of Species 6   || CoDeepNEAT DENSE Module | ID:   #122 | Fitness: 155.8334 | Units:   32 | Activ:   relu | Dropout: None\n",
            "\n",
            "\u001b[1m##############################################################################################################################################\u001b[0m\n",
            "\n",
            "Backed up generation 13 of the CoDeepNEAT evolutionary run to file: /content/tfne_state_backups/tfne_state_backup_2021-Sep-27_20-39-27/tfne_state_backup_gen_13.json\n",
            "\n",
            "Evaluating 72 genomes in generation 14...\n",
            "[========================================] 72/72 Genomes | Genome ID 1088 achieved fitness of 175.0\n",
            "\n",
            "\n",
            "\u001b[1m############################################################  Population Summary  ############################################################\n",
            "\n",
            "Generation:   14  ||  Best Genome Fitness: 209.93667  ||  Avg Blueprint Fitness: 177.0881  ||  Avg Module Fitness: 159.1955\u001b[0m\n",
            "Best Genome: CoDeepNEAT Genome | ID:    262 | Fitness: 209.93667 | Blueprint ID:     27 | Module Species: {1} | Optimizer:    sgd | Origin Gen:    3\n",
            "\n",
            "\u001b[1mBlueprint Species       || Blueprint Species Avg Fitness       || Blueprint Species Size\u001b[0m\n",
            "     1                  ||  171.875                            ||        3\n",
            "Best BP of Species 1    || CoDeepNEAT Blueprint | ID:    #35 | Fitness: 171.875 | Nodes:    2 | Module Species: {1} | Optimizer: sgd\n",
            "     2                  || 176.0579                            ||        2\n",
            "Best BP of Species 2    || CoDeepNEAT Blueprint | ID:    #47 | Fitness: 180.2408 | Nodes:    4 | Module Species: {1, 2, 4} | Optimizer: sgd\n",
            "     3                  ||  181.513                            ||        4\n",
            "Best BP of Species 3    || CoDeepNEAT Blueprint | ID:    #74 | Fitness: 183.4058 | Nodes:    4 | Module Species: {1, 2} | Optimizer: sgd\n",
            "\n",
            "\u001b[1mModule Species          || Module Species Avg Fitness          || Module Species Size\u001b[0m\n",
            "     1                  || 177.6971                            ||        5\n",
            "Best Mod of Species 1   || CoDeepNEAT DENSE Module | ID:   #196 | Fitness: 182.5337 | Units:   32 | Activ:   tanh | Dropout: None\n",
            "     2                  || 181.7095                            ||        4\n",
            "Best Mod of Species 2   || CoDeepNEAT DENSE Module | ID:   #198 | Fitness: 186.0744 | Units:   76 | Activ:   tanh | Dropout: None\n",
            "     3                  || 181.9182                            ||        4\n",
            "Best Mod of Species 3   || CoDeepNEAT DENSE Module | ID:   #195 | Fitness: 191.5983 | Units:   40 | Activ:   tanh | Dropout: None\n",
            "     4                  || 179.8967                            ||        5\n",
            "Best Mod of Species 4   || CoDeepNEAT DENSE Module | ID:   #181 | Fitness: 186.0833 | Units:   44 | Activ:   relu | Dropout: None\n",
            "     5                  || 111.5741                            ||        4\n",
            "Best Mod of Species 5   || CoDeepNEAT DENSE Module | ID:   #135 | Fitness: 152.7778 | Units:   32 | Activ: softmax | Dropout: None\n",
            "     6                  ||  97.0371                            ||        3\n",
            "Best Mod of Species 6   || CoDeepNEAT DENSE Module | ID:   #122 | Fitness: 155.8334 | Units:   32 | Activ:   relu | Dropout: None\n",
            "\n",
            "\u001b[1m##############################################################################################################################################\u001b[0m\n",
            "\n",
            "Backed up generation 14 of the CoDeepNEAT evolutionary run to file: /content/tfne_state_backups/tfne_state_backup_2021-Sep-27_20-39-27/tfne_state_backup_gen_14.json\n",
            "\n",
            "Evaluating 80 genomes in generation 15...\n",
            "[========================================] 80/80 Genomes | Genome ID 1168 achieved fitness of 175.0\n",
            "\n",
            "\n",
            "\u001b[1m############################################################  Population Summary  ############################################################\n",
            "\n",
            "Generation:   15  ||  Best Genome Fitness: 209.93667  ||  Avg Blueprint Fitness: 177.4511  ||  Avg Module Fitness:  152.757\u001b[0m\n",
            "Best Genome: CoDeepNEAT Genome | ID:    262 | Fitness: 209.93667 | Blueprint ID:     27 | Module Species: {1} | Optimizer:    sgd | Origin Gen:    3\n",
            "\n",
            "\u001b[1mBlueprint Species       || Blueprint Species Avg Fitness       || Blueprint Species Size\u001b[0m\n",
            "     1                  || 173.3307                            ||        3\n",
            "Best BP of Species 1    || CoDeepNEAT Blueprint | ID:    #35 | Fitness: 176.2421 | Nodes:    2 | Module Species: {1} | Optimizer: sgd\n",
            "     2                  || 174.3764                            ||        3\n",
            "Best BP of Species 2    || CoDeepNEAT Blueprint | ID:    #47 | Fitness: 184.4237 | Nodes:    4 | Module Species: {1, 2, 4} | Optimizer: sgd\n",
            "     3                  || 182.8474                            ||        4\n",
            "Best BP of Species 3    || CoDeepNEAT Blueprint | ID:    #73 | Fitness: 184.3304 | Nodes:    5 | Module Species: {1, 2, 3, 4} | Optimizer: sgd\n",
            "\n",
            "\u001b[1mModule Species          || Module Species Avg Fitness          || Module Species Size\u001b[0m\n",
            "     1                  ||  177.775                            ||        4\n",
            "Best Mod of Species 1   || CoDeepNEAT DENSE Module | ID:   #186 | Fitness: 180.6884 | Units:   32 | Activ:   tanh | Dropout: None\n",
            "     2                  || 184.7907                            ||        4\n",
            "Best Mod of Species 2   || CoDeepNEAT DENSE Module | ID:   #191 | Fitness: 197.3178 | Units:   76 | Activ:   tanh | Dropout: None\n",
            "     3                  || 144.9762                            ||        5\n",
            "Best Mod of Species 3   || CoDeepNEAT DENSE Module | ID:   #195 | Fitness: 197.1311 | Units:   40 | Activ:   tanh | Dropout: None\n",
            "     4                  || 181.2747                            ||        5\n",
            "Best Mod of Species 4   || CoDeepNEAT DENSE Module | ID:    #96 | Fitness: 195.2929 | Units:   44 | Activ:   relu | Dropout: None\n",
            "     5                  || 111.5741                            ||        4\n",
            "Best Mod of Species 5   || CoDeepNEAT DENSE Module | ID:   #135 | Fitness: 152.7778 | Units:   32 | Activ: softmax | Dropout: None\n",
            "     6                  ||  97.0371                            ||        3\n",
            "Best Mod of Species 6   || CoDeepNEAT DENSE Module | ID:   #122 | Fitness: 155.8334 | Units:   32 | Activ:   relu | Dropout: None\n",
            "\n",
            "\u001b[1m##############################################################################################################################################\u001b[0m\n",
            "\n",
            "Backed up generation 15 of the CoDeepNEAT evolutionary run to file: /content/tfne_state_backups/tfne_state_backup_2021-Sep-27_20-39-27/tfne_state_backup_gen_15.json\n",
            "Population reached specified maximum number of generations.\n",
            "Exiting evolutionary training loop...\n",
            "\n",
            "> Mejor RNA generada por la evolución:\n",
            "\n",
            "CoDeepNEAT Genome | ID:    262 | Fitness: 209.93667 | Blueprint ID:     27 | Module Species: {1} | Optimizer:    sgd | Origin Gen:    3\n",
            "\n",
            "\n",
            "Saved CoDeepNEAT genome (ID: 262) to file: ./best_genome_genotype_ori/genome_262_genotype.json\n",
            "INFO:tensorflow:Assets written to: ./best_genome_model_ori/assets\n",
            "\n",
            "\n",
            "\n",
            "> Re-entrenando mejor RNA por 350 épocas...\n",
            "\n",
            "> Probando Genome #262:\n",
            "  Aptitud lograda:\t175.0\n",
            "\n",
            "\n",
            "\n",
            "Saved CoDeepNEAT genome (ID: 262) to file: ./best_genome_genotype/genome_262_genotype.json\n",
            "INFO:tensorflow:Assets written to: ./best_genome_model/assets\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "s30bw_ff7Zj4"
      },
      "source": [
        "4) Evaluar el modelo de la RNA generada:"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "dDr_EuIkNnI-",
        "cellView": "form",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "5924c406-100a-4e8d-c349-fa7b1e697c67"
      },
      "source": [
        "#@title Cargar modelo de mejor genoma de RNA generada\n",
        "\n",
        "usar_modelo = \"Re-Entrenado\" #@param [\"Original\", \"Re-Entrenado\"]\n",
        "\n",
        "if usar_modelo == \"Original\":\n",
        "  best_model_dir = best_model_ori_dir\n",
        "else:\n",
        "  best_model_dir= best_model_new_dir\n",
        "\n",
        "modelRNA = tf.keras.models.load_model(best_model_dir)\n",
        "print(\"\\nModelo recuperado de \", best_model_dir)\n",
        "\n",
        "print(\"\\n\")\n",
        "modelRNA.summary()\n",
        "print(\"\\n\")\n"
      ],
      "execution_count": 13,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "Modelo recuperado de  ./best_genome_model/\n",
            "\n",
            "\n",
            "Model: \"model\"\n",
            "_________________________________________________________________\n",
            "Layer (type)                 Output Shape              Param #   \n",
            "=================================================================\n",
            "input_1 (InputLayer)         [(None, 15)]              0         \n",
            "_________________________________________________________________\n",
            "dense (Dense)                (None, 32)                512       \n",
            "_________________________________________________________________\n",
            "dense_1 (Dense)              (None, 8)                 264       \n",
            "=================================================================\n",
            "Total params: 776\n",
            "Trainable params: 776\n",
            "Non-trainable params: 0\n",
            "_________________________________________________________________\n",
            "\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "kVDw0-o-Oca0",
        "cellView": "form",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "cc25ca06-83ee-444c-9d44-2555e80f23eb"
      },
      "source": [
        "#@title Probar RNA con datos de entrenamiento\n",
        "# función auxiliar para probar el modelo entrenado en detalle\n",
        "def probarModelo(x, y, clases_map):\n",
        "\n",
        "    # procesa los datos de prueba con el modelo \n",
        "    predClass = modelRNA.predict(x)\n",
        "\n",
        "    # muestra los resultados con las imágenes \n",
        "    umbralClas = 0.5\n",
        "    classPreds = []\n",
        "    classReal = []\n",
        "    for i in range(len(x)):\n",
        "\n",
        "        # prepara salida\n",
        "        clReal = clases_map[ y[i] ] \n",
        "        \n",
        "        ## determina clase predecida de acuerdo a la que tiene mayor valor\n",
        "        idclPred = int( np.argmax(predClass[i], axis=0) )\n",
        "\n",
        "        if idclPred<0 or idclPred>=len(clases_map):\n",
        "            clPred = \"CLASE \" + str(idclPred) + \" INVÁLIDA\"\n",
        "        else:      \n",
        "            clPred = clases_map[ idclPred ]\n",
        "\n",
        "        classReal.append( clReal )\n",
        "        classPreds.append( clPred )\n",
        "        \n",
        "        strTitulo = 'Real: ' + clReal + ' / RNA: ' \n",
        "        strTitulo = strTitulo + clPred + ' (' + str( idclPred ) +')'   \n",
        "        strTitulo = strTitulo + \": \" + (\"ok\" if (clPred==clReal) else \"error!\")\n",
        "\n",
        "        # muestra comparación con la imagen\n",
        "        ##print(strTitulo)\n",
        "\n",
        "    # muestra reporte de clasificación\n",
        "    print(\"\\n Reporte de Clasificación: \")\n",
        "    print(classification_report(classReal, classPreds))\n",
        "\n",
        "    # muestra matriz de confusion\n",
        "    print('\\nMatriz de Confusión: ')\n",
        "    cm = confusion_matrix(classReal, classPreds, labels=clases_map)\n",
        "    cmtx = pd.DataFrame(\n",
        "        cm, \n",
        "        index=['r:{:}'.format(x) for x in clases_map], \n",
        "        columns=['p:{:}'.format(x) for x in clases_map]\n",
        "      )\n",
        "    # agrega para poder mostrar la matrix de confusión completa\n",
        "    pd.options.display.max_rows = 100\n",
        "    pd.options.display.max_columns = 100\n",
        "    print(cmtx)\n",
        "    print(\"\\n\")\n",
        "\n",
        "\n",
        "# prueba con los datos de entrenamiento\n",
        "print(\"*** Resultados con datos de Entrenamiento: \")\n",
        "probarModelo(x_train, y_train, CLASES)\n"
      ],
      "execution_count": 14,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "*** Resultados con datos de Entrenamiento: \n",
            "\n",
            " Reporte de Clasificación: \n",
            "              precision    recall  f1-score   support\n",
            "\n",
            "     ANFIBIO       1.00      1.00      1.00         3\n",
            "         AVE       0.86      1.00      0.92         6\n",
            "     INSECTO       1.00      1.00      1.00         4\n",
            "INVERTEBRADO       1.00      0.80      0.89         5\n",
            "    MAMIFERO       1.00      1.00      1.00        10\n",
            "         PEZ       1.00      1.00      1.00         3\n",
            "      REPTIL       1.00      1.00      1.00         3\n",
            "\n",
            "    accuracy                           0.97        34\n",
            "   macro avg       0.98      0.97      0.97        34\n",
            "weighted avg       0.97      0.97      0.97        34\n",
            "\n",
            "\n",
            "Matriz de Confusión: \n",
            "                p:na  p:MAMIFERO  p:AVE  p:REPTIL  p:PEZ  p:ANFIBIO  \\\n",
            "r:na               0           0      0         0      0          0   \n",
            "r:MAMIFERO         0          10      0         0      0          0   \n",
            "r:AVE              0           0      6         0      0          0   \n",
            "r:REPTIL           0           0      0         3      0          0   \n",
            "r:PEZ              0           0      0         0      3          0   \n",
            "r:ANFIBIO          0           0      0         0      0          3   \n",
            "r:INSECTO          0           0      0         0      0          0   \n",
            "r:INVERTEBRADO     0           0      1         0      0          0   \n",
            "\n",
            "                p:INSECTO  p:INVERTEBRADO  \n",
            "r:na                    0               0  \n",
            "r:MAMIFERO              0               0  \n",
            "r:AVE                   0               0  \n",
            "r:REPTIL                0               0  \n",
            "r:PEZ                   0               0  \n",
            "r:ANFIBIO               0               0  \n",
            "r:INSECTO               4               0  \n",
            "r:INVERTEBRADO          0               4  \n",
            "\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "OoQ7iOSdQF30",
        "cellView": "form",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "35f2d137-340c-4cfc-897e-f23beb29f956"
      },
      "source": [
        "#@title Probar RNA con datos de prueba\n",
        "\n",
        "print(\"*** Resultados con datos de Prueba: \")\n",
        "probarModelo(x_test, y_test, CLASES)"
      ],
      "execution_count": 15,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "*** Resultados con datos de Prueba: \n",
            "\n",
            " Reporte de Clasificación: \n",
            "              precision    recall  f1-score   support\n",
            "\n",
            "     ANFIBIO       1.00      1.00      1.00         1\n",
            "         AVE       1.00      1.00      1.00         2\n",
            "     INSECTO       1.00      1.00      1.00         2\n",
            "INVERTEBRADO       1.00      1.00      1.00         3\n",
            "    MAMIFERO       1.00      1.00      1.00         5\n",
            "         PEZ       1.00      1.00      1.00         1\n",
            "      REPTIL       1.00      1.00      1.00         2\n",
            "\n",
            "    accuracy                           1.00        16\n",
            "   macro avg       1.00      1.00      1.00        16\n",
            "weighted avg       1.00      1.00      1.00        16\n",
            "\n",
            "\n",
            "Matriz de Confusión: \n",
            "                p:na  p:MAMIFERO  p:AVE  p:REPTIL  p:PEZ  p:ANFIBIO  \\\n",
            "r:na               0           0      0         0      0          0   \n",
            "r:MAMIFERO         0           5      0         0      0          0   \n",
            "r:AVE              0           0      2         0      0          0   \n",
            "r:REPTIL           0           0      0         2      0          0   \n",
            "r:PEZ              0           0      0         0      1          0   \n",
            "r:ANFIBIO          0           0      0         0      0          1   \n",
            "r:INSECTO          0           0      0         0      0          0   \n",
            "r:INVERTEBRADO     0           0      0         0      0          0   \n",
            "\n",
            "                p:INSECTO  p:INVERTEBRADO  \n",
            "r:na                    0               0  \n",
            "r:MAMIFERO              0               0  \n",
            "r:AVE                   0               0  \n",
            "r:REPTIL                0               0  \n",
            "r:PEZ                   0               0  \n",
            "r:ANFIBIO               0               0  \n",
            "r:INSECTO               2               0  \n",
            "r:INVERTEBRADO          0               3  \n",
            "\n",
            "\n"
          ]
        }
      ]
    }
  ]
}